# Test
Bug 12 - patch summary with no bug number

Some bogus patch description
diff --git a/tests/test5.cpp b/tests/test5.cpp
new file mode 100644
--- /dev/null
+++ b/tests/test5.cpp
@@ -0,0 +1,24 @@
+// License bogus
+
+// runtime/virtual
+class ShouldHaveVirtualDes {
+  virtual foo();
+};
+
+int main() {
+  // runtime/memset
+  memset(blah, sizeof(blah), 0);
+
+  // runtime/rtti
+  dynamic_cast<Derived*>(obj);
+
+  // runtime/sizeof
+  int varname = 0;
+  int mySize = sizeof(int);
+
+  // runtime/threadsafe_fn
+  getpwuid();
+  strtok();
+
+  return 0;
+}
