--- mozilla/security/coreconf/location.mk	21 Jul 2005 23:48:08 -0000	1.10
+++ mozilla/security/coreconf/location.mk	1 Feb 2006 22:01:10 -0000
@@ -70,9 +70,17 @@
 ifdef NSPR_INCLUDE_DIR
     INCLUDES += -I$(NSPR_INCLUDE_DIR)
 endif
 
 ifndef NSPR_LIB_DIR
     NSPR_LIB_DIR = $(DIST)/lib
 endif
 
+ifdef NSS_INCLUDE_DIR
+    INCLUDES += -I$(NSS_INCLUDE_DIR)
+endif
+
+ifndef NSS_LIB_DIR
+    NSS_LIB_DIR = $(DIST)/lib
+endif
+
 MK_LOCATION = included
