error[E0308]: mismatched types
  --> tests/ui/cfg/cfg_attr-type-mismatch.rs:19:27
   |
19 |     let _: Pin<&mut u8> = x.inner; //~ ERROR E0308
   |            ------------   ^^^^^^^ expected struct `Pin`, found `&mut u8`
   |            |
   |            expected due to this
   |
   = note:         expected struct `Pin<&mut u8>`
           found mutable reference `&mut u8`

error[E0308]: mismatched types
  --> tests/ui/cfg/cfg_attr-type-mismatch.rs:23:22
   |
23 |     let _: &mut u8 = x.inner; //~ ERROR E0308
   |            -------   ^^^^^^^
   |            |         |
   |            |         expected `&mut u8`, found struct `Pin`
   |            |         help: consider mutably borrowing here: `&mut x.inner`
   |            expected due to this
   |
   = note: expected mutable reference `&mut u8`
                         found struct `Pin<&mut u8>`
