diff --git a/media/libtheora/lib/state.c b/media/libtheora/lib/state.c
--- a/media/libtheora/lib/state.c
+++ b/media/libtheora/lib/state.c
@@ -87,17 +87,17 @@ static void oc_sb_create_plane_mapping(o
       int       quadi;
       int       i;
       /*Figure out how many rows of blocks in this super block lie within the
          image.*/
       jmax=_hfrags-x;
       if(jmax>4)jmax=4;
       else if(jmax<=0)break;
       /*By default, set all fragment indices to -1.*/
-      memset(_sb_maps[sbi][0],0xFF,sizeof(_sb_maps[sbi]));
+      memset(_sb_maps[sbi],0xFF,sizeof(_sb_maps[sbi]));
       /*Fill in the fragment map for this super block.*/
       xfrag=yfrag+x;
       for(i=0;i<imax;i++){
         int j;
         for(j=0;j<jmax;j++){
           _sb_maps[sbi][SB_MAP[i][j][0]][SB_MAP[i][j][1]]=xfrag+j;
         }
         xfrag+=_hfrags;
