diff --git a/modules/ResetProfile.sys.mjs b/modules/ResetProfile.sys.mjs
index e1b8ae6..9b4c7e4 100755
--- ./modules/ResetProfile.sys.mjs
+++ ./modules/ResetProfile.sys.mjs
@@ -40,7 +40,7 @@ export var ResetProfile = {
     // Reset is only supported if the self-migrator used for reset exists.
     if (
       !lazy.MigrationUtils ||
-      !lazy.MigrationUtils.migratorExists(MOZ_APP_NAME)
+      !lazy.MigrationUtils.migratorExists("firefox")
     ) {
       return false;
     }
