diff --git a/extensions/spellcheck/hunspell/src/filemgr.hxx b/extensions/spellcheck/hunspell/src/filemgr.hxx
--- a/extensions/spellcheck/hunspell/src/filemgr.hxx
+++ b/extensions/spellcheck/hunspell/src/filemgr.hxx
@@ -67,32 +67,11 @@
  * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
  * SUCH DAMAGE.
  */
 
 /* file manager class - read lines of files [filename] OR [filename.hz] */
 #ifndef FILEMGR_HXX_
 #define FILEMGR_HXX_
 
-#include "hunzip.hxx"
-#include <stdio.h>
-#include <string>
-#include <fstream>
-
-class FileMgr {
- private:
-  FileMgr(const FileMgr&);
-  FileMgr& operator=(const FileMgr&);
+#include "mozHunspellRLBoxSandbox.h"
 
- protected:
-  std::ifstream fin;
-  Hunzip* hin;
-  char in[BUFSIZE + 50];  // input buffer
-  int fail(const char* err, const char* par);
-  int linenum;
-
- public:
-  FileMgr(const char* filename, const char* key = NULL);
-  ~FileMgr();
-  bool getline(std::string&);
-  int getlinenum();
-};
 #endif
