From 2c1d5189aedc7f90b603807a36a57254dc24471c Mon Sep 17 00:00:00 2001
From: Dan Glastonbury <dglastonbury@mozilla.com>
Date: Fri, 19 Jan 2018 11:56:58 +1000
Subject: gecko: Change paths to vendored crates.


diff --git a/media/audioipc/Cargo.toml b/media/audioipc/Cargo.toml
index ede6064..d0a1979 100644
--- a/media/audioipc/Cargo.toml
+++ b/media/audioipc/Cargo.toml
@@ -1,2 +1,2 @@
 [workspace]
-members = ["audioipc", "client", "server", "ipctest"]
+members = ["audioipc", "client", "server"]

