From 2c1d5189aedc7f90b603807a36a57254dc24471c Mon Sep 17 00:00:00 2001
From: Dan Glastonbury <dglastonbury@mozilla.com>
Date: Fri, 19 Jan 2018 11:56:58 +1000
Subject: gecko: Change paths to vendored crates.


diff --git a/media/audioipc/Cargo.toml b/media/audioipc/Cargo.toml
index ede6064..d0a1979 100644
--- a/media/audioipc/Cargo.toml
+++ b/media/audioipc/Cargo.toml
@@ -1,2 +1,2 @@
 [workspace]
-members = ["audioipc", "client", "server", "ipctest"]
+members = ["audioipc", "client", "server"]

Fix deprecated use of config due to bincode update:
--- a/media/audioipc/audioipc/src/codec.rs
+++ b/media/audioipc/audioipc/src/codec.rs
@@ -5,7 +5,7 @@
 
 //! `Encoder`s and `Decoder`s from items to/from `BytesMut` buffers.
 
-use bincode::{self, deserialize, serialized_size};
+use bincode::{self, deserialize, serialized_size, Options};
 use bytes::{BufMut, ByteOrder, BytesMut, LittleEndian};
 use serde::de::DeserializeOwned;
 use serde::ser::Serialize;
@@ -170,8 +170,8 @@ where
 
         buf.put_u32_le(encoded_len as u32);
 
-        if let Err(e) = bincode::config()
-            .limit(encoded_len)
+        if let Err(e) = bincode::options()
+            .with_limit(encoded_len)
             .serialize_into::<_, Self::In>(&mut buf.writer(), &item)
         {
             match *e {
