diff --git a/media/libav/libavutil/internal.h b/media/libav/libavutil/internal.h
index aed9925..03ee543 100644
--- a/media/libav/libavutil/internal.h
+++ b/media/libav/libavutil/internal.h
@@ -134,7 +134,7 @@
 
 #include "libm.h"
 
-#if defined(_MSC_VER)
+#if defined(_MSC_VER) && _MSC_VER < 1800
 #pragma comment(linker, "/include:"EXTERN_PREFIX"avpriv_strtod")
 #pragma comment(linker, "/include:"EXTERN_PREFIX"avpriv_snprintf")
 #endif
