--- ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo.c	2005-05-11 19:59:13.000000000 -0700
+++ cairo/src/cairo.c	2005-06-01 23:07:50.000000000 -0700
@@ -115,12 +115,6 @@
 
     _cairo_path_fixed_init (&cr->path);
 
-    if (target == NULL) {
-	cr->gstate = NULL;
-	cr->status = CAIRO_STATUS_NULL_POINTER;
-	return cr;
-    }
-
     cr->gstate = _cairo_gstate_create (target);
     if (cr->gstate == NULL)
 	cr->status = CAIRO_STATUS_NO_MEMORY;
