diff -r 0e7cd985a7f0 gfx/cairo/cairo/src/cairo-traps.c
--- a/gfx/cairo/cairo/src/cairo-traps.c	Fri Jun 08 18:09:53 2007 -0700
+++ b/gfx/cairo/cairo/src/cairo-traps.c	Fri Jun 08 18:10:48 2007 -0700
@@ -599,6 +599,12 @@ _cairo_traps_extract_region (cairo_traps
 {
     int i;
 
+    /* Bail early if we have lots of traps, until we fix the code
+     * below to not use Union()
+     */
+    if (traps->num_traps > 200)
+        return CAIRO_INT_STATUS_UNSUPPORTED;
+
     for (i = 0; i < traps->num_traps; i++)
 	if (!(traps->traps[i].left.p1.x == traps->traps[i].left.p2.x
 	      && traps->traps[i].right.p1.x == traps->traps[i].right.p2.x
