commit c32b57b9ada7a57ec20648629ecb83de5688682a
Author: Jeff Muizelaar <jmuizelaar@mozilla.com>
Date:   Mon Mar 23 11:28:12 2009 -0400

    shrink cache size

diff --git a/src/cairo-scaled-font.c b/src/cairo-scaled-font.c
index 249ab6c..aa7fc11 100644
--- a/src/cairo-scaled-font.c
+++ b/src/cairo-scaled-font.c
@@ -63,7 +63,7 @@
  */
 
 /* XXX: This number is arbitrary---we've never done any measurement of this. */
-#define MAX_GLYPH_PAGES_CACHED 512
+#define MAX_GLYPH_PAGES_CACHED 256
 static cairo_cache_t *cairo_scaled_glyph_page_cache;
 
 #define CAIRO_SCALED_GLYPH_PAGE_SIZE 32
