# HG changeset patch
# Parent 2563fa2763b0ea83394e785340afa4c564ceab57
diff -r 2563fa2763b0 -r 9ab15e95a354 gfx/cairo/cairo/src/cairo-qt-surface.cpp
--- a/gfx/cairo/cairo/src/cairo-qt-surface.cpp	Thu Apr 29 06:55:11 2010 +0300
+++ b/gfx/cairo/cairo/src/cairo-qt-surface.cpp	Thu Apr 29 06:55:51 2010 +0300
@@ -204,6 +204,7 @@
     case CAIRO_OPERATOR_HSL_LUMINOSITY:
 	ASSERT_NOT_REACHED;
     }
+    return QPainter::CompositionMode_Source;
 }
 
 static bool
@@ -668,7 +669,6 @@
 static cairo_status_t
 _cairo_qt_surface_clone_similar (void *abstract_surface,
 				 cairo_surface_t *src,
-				 cairo_content_t  content,
 				 int              src_x,
 				 int              src_y,
 				 int              width,
