diff -ur ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo-atsui.h cairo/src/cairo-atsui.h
--- ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo-atsui.h	2005-05-14 13:20:32.000000000 -0700
+++ cairo/src/cairo-atsui.h	2005-06-01 22:36:04.000000000 -0700
@@ -38,7 +38,7 @@
 
 #include <cairo.h>
 
-#if CAIRO_HAS_ATSUI_FONT
+#ifdef CAIRO_HAS_ATSUI_FONT
 
 /* ATSUI platform-specific font interface */
 
diff -ur ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo-features.h.in cairo/src/cairo-features.h.in
--- ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo-features.h.in	2005-05-07 09:54:50.000000000 -0700
+++ cairo/src/cairo-features.h.in	2005-06-01 19:46:42.000000000 -0700
@@ -37,26 +37,28 @@
 #ifndef CAIRO_FEATURES_H
 #define CAIRO_FEATURES_H
 
-@PS_SURFACE_FEATURE@
+#define @PS_SURFACE_FEATURE@
 
-@PDF_SURFACE_FEATURE@
+#define @PDF_SURFACE_FEATURE@
 
-@XLIB_SURFACE_FEATURE@
+#define @PNG_SURFACE_FEATURE@
 
-@QUARTZ_SURFACE_FEATURE@
+#define @XLIB_SURFACE_FEATURE@
 
-@XCB_SURFACE_FEATURE@
+#define @QUARTZ_SURFACE_FEATURE@
 
-@WIN32_SURFACE_FEATURE@
+#define @XCB_SURFACE_FEATURE@
 
-@GLITZ_SURFACE_FEATURE@
+#define @WIN32_SURFACE_FEATURE@
 
-@FT_FONT_FEATURE@
+#define @GLITZ_SURFACE_FEATURE@
 
-@WIN32_FONT_FEATURE@
+#define @FT_FONT_FEATURE@
 
-@ATSUI_FONT_FEATURE@
+#define @WIN32_FONT_FEATURE@
 
-@PNG_FUNCTIONS_FEATURE@
+#define @ATSUI_FONT_FEATURE@
+
+#define @SANITY_CHECKING_FEATURE@
 
 #endif
diff -ur ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo-ft-private.h cairo/src/cairo-ft-private.h
--- ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo-ft-private.h	2005-05-06 21:55:21.000000000 -0700
+++ cairo/src/cairo-ft-private.h	2005-06-01 22:36:15.000000000 -0700
@@ -40,7 +40,7 @@
 #include <cairo-ft.h>
 #include <cairoint.h>
 
-#if CAIRO_HAS_FT_FONT
+#ifdef CAIRO_HAS_FT_FONT
 
 CAIRO_BEGIN_DECLS
 
diff -ur ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo-ft.h cairo/src/cairo-ft.h
--- ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo-ft.h	2005-05-14 13:20:32.000000000 -0700
+++ cairo/src/cairo-ft.h	2005-06-01 22:36:20.000000000 -0700
@@ -39,7 +39,7 @@
 
 #include <cairo.h>
 
-#if CAIRO_HAS_FT_FONT
+#ifdef CAIRO_HAS_FT_FONT
 
 /* Fontconfig/Freetype platform-specific font interface */
 
diff -ur ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo-glitz.h cairo/src/cairo-glitz.h
--- ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo-glitz.h	2005-05-14 13:20:32.000000000 -0700
+++ cairo/src/cairo-glitz.h	2005-06-01 22:36:30.000000000 -0700
@@ -39,7 +39,7 @@
 
 #include <cairo.h>
 
-#if CAIRO_HAS_GLITZ_SURFACE
+#ifdef CAIRO_HAS_GLITZ_SURFACE
 
 #include <glitz.h>
 
diff -ur ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo-gstate-private.h cairo/src/cairo-gstate-private.h
--- ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo-gstate-private.h	2005-05-06 21:55:23.000000000 -0700
+++ cairo/src/cairo-gstate-private.h	2005-06-01 22:19:22.000000000 -0700
@@ -52,6 +52,8 @@
     double *dash;
     int num_dashes;
     double dash_offset;
+    double max_dash_length;
+    double fraction_dash_lit;
 
     char *font_family; /* NULL means CAIRO_FONT_FAMILY_DEFAULT; */
     cairo_font_slant_t font_slant; 
diff -ur ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo-pdf.h cairo/src/cairo-pdf.h
--- ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo-pdf.h	2005-05-17 06:17:45.000000000 -0700
+++ cairo/src/cairo-pdf.h	2005-06-01 22:36:43.000000000 -0700
@@ -39,7 +39,7 @@
 
 #include <cairo.h>
 
-#if CAIRO_HAS_PDF_SURFACE
+#ifdef CAIRO_HAS_PDF_SURFACE
 
 CAIRO_BEGIN_DECLS
 
diff -ur ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo-ps.h cairo/src/cairo-ps.h
--- ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo-ps.h	2005-05-17 06:17:45.000000000 -0700
+++ cairo/src/cairo-ps.h	2005-06-01 22:38:24.000000000 -0700
@@ -39,7 +39,7 @@
 
 #include <cairo.h>
 
-#if CAIRO_HAS_PS_SURFACE
+#ifdef CAIRO_HAS_PS_SURFACE
 
 #include <stdio.h>
 
diff -ur ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo-quartz.h cairo/src/cairo-quartz.h
--- ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo-quartz.h	2005-05-14 13:20:32.000000000 -0700
+++ cairo/src/cairo-quartz.h	2005-06-01 22:36:53.000000000 -0700
@@ -39,7 +39,7 @@
 
 #include <cairo.h>
 
-#if CAIRO_HAS_QUARTZ_SURFACE
+#ifdef CAIRO_HAS_QUARTZ_SURFACE
 
 #include <Carbon/Carbon.h>
 
diff -ur ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo-wideint.h cairo/src/cairo-wideint.h
--- ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo-wideint.h	2005-05-10 14:55:37.000000000 -0700
+++ cairo/src/cairo-wideint.h	2005-06-01 22:07:43.000000000 -0700
@@ -45,7 +45,7 @@
 #elif HAVE_SYS_INT_TYPES_H
 # include <sys/int_types.h>
 #else
-#error Cannot find definitions for fixed-width integral types (uint8_t, uint32_t, etc.)
+# include "mozstdint.h"
 #endif
 
 /*
diff -ur ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo-win32.h cairo/src/cairo-win32.h
--- ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo-win32.h	2005-05-14 13:20:32.000000000 -0700
+++ cairo/src/cairo-win32.h	2005-06-01 22:37:00.000000000 -0700
@@ -38,7 +38,7 @@
 
 #include <cairo.h>
 
-#if CAIRO_HAS_WIN32_SURFACE
+#ifdef CAIRO_HAS_WIN32_SURFACE
 
 #include <windows.h>
 
diff -ur ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo-xcb-xrender.h cairo/src/cairo-xcb-xrender.h
--- ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo-xcb-xrender.h	2005-05-17 06:11:32.000000000 -0700
+++ cairo/src/cairo-xcb-xrender.h	2005-06-01 22:38:33.000000000 -0700
@@ -39,7 +39,7 @@
 
 #include <cairo.h>
 
-#if CAIRO_HAS_XCB_SURFACE
+#ifdef CAIRO_HAS_XCB_SURFACE
 
 #include <X11/XCB/xcb.h>
 #include <X11/XCB/render.h>
diff -ur ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo-xcb.h cairo/src/cairo-xcb.h
--- ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo-xcb.h	2005-05-17 06:17:45.000000000 -0700
+++ cairo/src/cairo-xcb.h	2005-06-01 22:38:38.000000000 -0700
@@ -39,7 +39,7 @@
 
 #include <cairo.h>
 
-#if CAIRO_HAS_XCB_SURFACE
+#ifdef CAIRO_HAS_XCB_SURFACE
 
 #include <X11/XCB/xcb.h>
 
diff -ur ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo-xlib-test.h cairo/src/cairo-xlib-test.h
--- ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo-xlib-test.h	2005-05-17 01:06:05.000000000 -0700
+++ cairo/src/cairo-xlib-test.h	2005-06-01 22:37:16.000000000 -0700
@@ -39,7 +39,7 @@
 
 #include <cairo.h>
 
-#if CAIRO_HAS_XLIB_SURFACE
+#ifdef CAIRO_HAS_XLIB_SURFACE
 
 #include <cairo-xlib.h>
 
Only in cairo/src: cairo-xlib-test.h~
diff -ur ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo-xlib-xrender.h cairo/src/cairo-xlib-xrender.h
--- ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo-xlib-xrender.h	2005-05-17 06:05:13.000000000 -0700
+++ cairo/src/cairo-xlib-xrender.h	2005-06-01 22:38:43.000000000 -0700
@@ -39,7 +39,7 @@
 
 #include <cairo.h>
 
-#if CAIRO_HAS_XLIB_SURFACE
+#ifdef CAIRO_HAS_XLIB_SURFACE
 
 #include <X11/extensions/Xrender.h>
 
Only in cairo/src: cairo-xlib-xrender.h~
diff -ur ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo-xlib.h cairo/src/cairo-xlib.h
--- ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo-xlib.h	2005-05-17 06:17:45.000000000 -0700
+++ cairo/src/cairo-xlib.h	2005-06-01 22:37:25.000000000 -0700
@@ -39,7 +39,7 @@
 
 #include <cairo.h>
 
-#if CAIRO_HAS_XLIB_SURFACE
+#ifdef CAIRO_HAS_XLIB_SURFACE
 
 #include <X11/Xlib.h>
 
Only in cairo/src: cairo-xlib.h~
diff -ur ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo.h cairo/src/cairo.h
--- ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairo.h	2005-05-17 08:29:48.000000000 -0700
+++ cairo/src/cairo.h	2005-06-01 22:39:03.000000000 -0700
@@ -935,7 +935,7 @@
 cairo_status_t
 cairo_surface_finish (cairo_surface_t *surface);
 
-#if CAIRO_HAS_PNG_FUNCTIONS
+#ifdef CAIRO_HAS_PNG_FUNCTIONS
 
 cairo_status_t
 cairo_surface_write_to_png (cairo_surface_t	*surface,
@@ -983,7 +983,7 @@
 int
 cairo_image_surface_get_height (cairo_surface_t *surface);
 
-#if CAIRO_HAS_PNG_FUNCTIONS
+#ifdef CAIRO_HAS_PNG_FUNCTIONS
 
 cairo_surface_t *
 cairo_image_surface_create_from_png (const char	*filename);
Only in cairo/src: cairo.h~
diff -ur ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairoint.h cairo/src/cairoint.h
--- ../../../../../cairo-cvs/dist/cairo-0.5.0/src/cairoint.h	2005-05-17 06:17:45.000000000 -0700
+++ cairo/src/cairoint.h	2005-06-01 22:38:09.000000000 -0700
@@ -546,19 +546,19 @@
 };
 
 /* concrete font backends */
-#if CAIRO_HAS_FT_FONT
+#ifdef CAIRO_HAS_FT_FONT
 
 extern const cairo_private struct _cairo_scaled_font_backend cairo_ft_scaled_font_backend;
 
 #endif
 
-#if CAIRO_HAS_WIN32_FONT
+#ifdef CAIRO_HAS_WIN32_FONT
 
 extern const cairo_private struct _cairo_scaled_font_backend cairo_win32_scaled_font_backend;
 
 #endif
 
-#if CAIRO_HAS_ATSUI_FONT
+#ifdef CAIRO_HAS_ATSUI_FONT
 
 extern const cairo_private struct _cairo_scaled_font_backend cairo_atsui_scaled_font_backend;
 
@@ -854,17 +854,17 @@
 #define CAIRO_FONT_SLANT_DEFAULT   CAIRO_FONT_SLANT_NORMAL
 #define CAIRO_FONT_WEIGHT_DEFAULT  CAIRO_FONT_WEIGHT_NORMAL
 
-#if   CAIRO_HAS_WIN32_FONT
+#if   defined(CAIRO_HAS_WIN32_FONT)
 
 #define CAIRO_FONT_FAMILY_DEFAULT "Arial"
 #define CAIRO_FONT_BACKEND_DEFAULT &cairo_win32_scaled_font_backend
 
-#elif CAIRO_HAS_ATSUI_FONT
+#elif defined(CAIRO_HAS_ATSUI_FONT)
 
 #define CAIRO_FONT_FAMILY_DEFAULT  "Monaco"
 #define CAIRO_FONT_BACKEND_DEFAULT &cairo_atsui_scaled_font_backend
 
-#elif CAIRO_HAS_FT_FONT
+#elif defined(CAIRO_HAS_FT_FONT)
 
 #define CAIRO_FONT_FAMILY_DEFAULT  "serif"
 #define CAIRO_FONT_BACKEND_DEFAULT &cairo_ft_scaled_font_backend

