commit 25fb70d4dc1e67553dba5973a79902e4cf81341f
Author: Jeff Muizelaar <jmuizelaar@mozilla.com>
Date:   Wed Apr 21 11:23:27 2010 -0400

    fix show glyphs clipping

diff --git a/src/cairo-win32-surface.c b/src/cairo-win32-surface.c
index c10e134..ff823a3 100644
--- a/src/cairo-win32-surface.c
+++ b/src/cairo-win32-surface.c
@@ -1603,6 +1603,8 @@ _cairo_win32_surface_show_glyphs (void			*surface,
 
 		_cairo_win32_surface_set_clip_region (surface, clip_region);
 	    }
+	} else {
+	    _cairo_win32_surface_set_clip_region (surface, NULL);
 	}
 
 	solid_pattern = (cairo_solid_pattern_t *)source;
