Index: configure.in
===================================================================
RCS file: /cvsroot/mozilla/gfx/cairo/cairo/configure.in,v
retrieving revision 1.1
diff -u -8 -p -u -r1.1 configure.in
--- configure.in	24 Feb 2005 20:08:53 -0000	1.1
+++ configure.in	8 Mar 2005 18:12:11 -0000
@@ -367,8 +367,16 @@ echo "  PDF: $use_pdf"
 echo "  PNG: $use_png"
 echo "  glitz: $use_glitz"
 echo ""
 echo "and the following font backends:"
 echo "  freetype: $use_freetype"
 echo "  atsui: $use_atsui"
 echo ""
 
+if test x"$use_freetype" != "xyes" && \
+   test x"$use_atsui"    != "xyes" ; then
+
+   AC_MSG_ERROR([Cairo requires at least one font backend.
+                  Please install freetype and fontconfig, then try again:
+                  http://freetype.org/  http://fontconfig.org/
+                ])
+fi
