commit d14b8c688f4cbe6c9289955d9ce7257c07869f9e
Author: Jeff Muizelaar <jmuizelaar@mozilla.com>
Date:   Wed May 6 10:19:04 2009 -0400

    create getter for component alpha

diff --git a/gfx/cairo/libpixman/src/pixman-image.c b/gfx/cairo/libpixman/src/pixman-image.c
index e80c479..2c49f92 100644
--- a/gfx/cairo/libpixman/src/pixman-image.c
+++ b/gfx/cairo/libpixman/src/pixman-image.c
@@ -563,6 +563,11 @@ pixman_image_set_component_alpha   (pixman_image_t       *image,
     image->common.component_alpha = component_alpha;
 }
 
+PIXMAN_EXPORT pixman_bool_t
+pixman_image_get_component_alpha   (pixman_image_t       *image)
+{
+    return image->common.component_alpha;
+}
 
 PIXMAN_EXPORT void
 pixman_image_set_accessors (pixman_image_t             *image,
diff --git a/gfx/cairo/libpixman/src/pixman.h b/gfx/cairo/libpixman/src/pixman.h
index 49c39d5..522a866 100644
--- a/gfx/cairo/libpixman/src/pixman.h
+++ b/gfx/cairo/libpixman/src/pixman.h
@@ -767,6 +767,7 @@ void            pixman_image_set_alpha_map           (pixman_image_t
 						      int16_t                       y);
 void            pixman_image_set_component_alpha     (pixman_image_t               *image,
 						      pixman_bool_t                 component_alpha);
+pixman_bool_t   pixman_image_get_component_alpha     (pixman_image_t               *image);
 void		pixman_image_set_accessors	     (pixman_image_t		   *image,
 						      pixman_read_memory_func_t	    read_func,
 						      pixman_write_memory_func_t    write_func);
