commit df2b22c8c6677d531194579c82a55e855adff706
Author: Jeff Muizelaar <jmuizelaar@mozilla.com>
Date:   Thu Apr 8 15:43:41 2010 -0400

    Propagate is_clear flag to quartz_image_surface

diff --git a/src/cairo-quartz-image-surface.c b/src/cairo-quartz-image-surface.c
index 5a624eb..b8809d5 100644
--- a/src/cairo-quartz-image-surface.c
+++ b/src/cairo-quartz-image-surface.c
@@ -146,6 +146,8 @@ _cairo_quartz_image_surface_flush (void *asurface)
     surface->image = newImage;
     CGImageRelease (oldImage);
 
+    surface->base.is_clear = surface->imageSurface->base.is_clear;
+
     return CAIRO_STATUS_SUCCESS;
 }
 
@@ -267,6 +269,8 @@ cairo_quartz_image_surface_create (cairo_surface_t *surface)
     qisurf->image = image;
     qisurf->imageSurface = image_surface;
 
+    qisurf->base.is_clear = image_surface->base.is_clear;
+
     return &qisurf->base;
 }
 
