--- libpixman-0.1.4/src/pixman.h	Fri Mar  4 07:25:30 2005
+++ libpixman/src/pixman.h	Mon Apr 18 13:52:47 2005
@@ -54,7 +54,7 @@
 SOFTWARE.
 
 ******************************************************************/
-/* $Id: stdint.diff,v 1.4 2005/04/19 17:19:43 bryner%brianryner.com Exp $ */
+/* $Id: stdint.diff,v 1.4 2005/04/19 17:19:43 bryner%brianryner.com Exp $ */
 
 /* libic.h */
 
@@ -81,14 +81,14 @@
  */
 
 
-#if defined (__SVR4) && defined (__sun)
+#if HAVE_STDINT_H
+# include <stdint.h>
+#elif HAVE_INTTYPES_H
+# include <inttypes.h>
+#elif HAVE_SYS_INT_TYPES_H
 # include <sys/int_types.h>
 #else
-# if defined (__OpenBSD__)
-#  include <inttypes.h>
-# else 
-#  include <stdint.h>
-# endif
+# include "mozstdint.h"
 #endif
 
 
--- cairo-0.4.0/src/cairo-wideint.h	Tue Feb 22 11:24:50 2005
+++ cairo/src/cairo-wideint.h	Mon Apr 18 13:53:56 2005
@@ -1,5 +1,5 @@
 /*
- * $Id: stdint.diff,v 1.4 2005/04/19 17:19:43 bryner%brianryner.com Exp $
+ * $Id: stdint.diff,v 1.4 2005/04/19 17:19:43 bryner%brianryner.com Exp $
  *
  * Copyright © 2004 Keith Packard
  *
@@ -38,7 +38,15 @@
 #ifndef CAIRO_WIDEINT_H
 #define CAIRO_WIDEINT_H
 
-#include <stdint.h>
+#if HAVE_STDINT_H
+# include <stdint.h>
+#elif HAVE_INTTYPES_H
+# include <inttypes.h>
+#elif HAVE_SYS_INT_TYPES_H
+# include <sys/int_types.h>
+#else
+# include "mozstdint.h"
+#endif
 
 /*
  * 64-bit datatypes.  Two separate implementations, one using
--- cairo-0.4.0/src/cairoint.h	Fri Mar  4 09:57:54 2005
+++ cairo/src/cairoint.h	Mon Apr 18 13:53:06 2005
@@ -54,7 +54,16 @@
 #include <string.h>
 #include <math.h>
 #include <limits.h>
-#include <stdint.h>
+
+#if HAVE_STDINT_H
+# include <stdint.h>
+#elif HAVE_INTTYPES_H
+# include <inttypes.h>
+#elif HAVE_SYS_INT_TYPES_H
+# include <sys/int_types.h>
+#else
+# include "mozstdint.h"
+#endif
 
 #include "cairo.h"
 
