--- cairo-0.4.0/src/cairo-wideint.h	2005-02-22 13:24:50.000000000 -0600
+++ cairo/src/cairo-wideint.h	2005-03-23 13:00:28.436158473 -0600
@@ -38,7 +38,15 @@
 #ifndef CAIRO_WIDEINT_H
 #define CAIRO_WIDEINT_H
 
-#include <stdint.h>
+#if defined (__SVR4) && defined (__sun)
+# include <sys/int_types.h>
+#else
+# if defined (__OpenBSD__)
+#  include <inttypes.h>
+# else 
+#  include <stdint.h>
+# endif
+#endif
 
 /*
  * 64-bit datatypes.  Two separate implementations, one using
--- cairo-0.4.0/src/cairoint.h	2005-03-04 11:57:54.000000000 -0600
+++ cairo/src/cairoint.h	2005-03-23 13:00:45.667289010 -0600
@@ -54,7 +54,16 @@
 #include <string.h>
 #include <math.h>
 #include <limits.h>
-#include <stdint.h>
+
+#if defined (__SVR4) && defined (__sun)
+# include <sys/int_types.h>
+#else
+# if defined (__OpenBSD__)
+#  include <inttypes.h>
+# else 
+#  include <stdint.h>
+# endif
+#endif
 
 #include "cairo.h"
 
