--- cairo-0.4.0/Makefile.am	2005-01-27 13:42:51.000000000 -0600
+++ cairo/Makefile.am	2005-03-23 11:13:34.000000000 -0600
@@ -1,4 +1,4 @@
-SUBDIRS = src test doc
+SUBDIRS = src
 
 EXTRA_DIST = \
 	COPYING \
--- cairo-0.4.0/Makefile.in	2005-03-08 19:43:54.000000000 -0600
+++ cairo/Makefile.in	2005-03-23 11:13:22.000000000 -0600
@@ -250,7 +250,7 @@
 sharedstatedir = @sharedstatedir@
 sysconfdir = @sysconfdir@
 target_alias = @target_alias@
-SUBDIRS = src test doc
+SUBDIRS = src
 EXTRA_DIST = \
 	COPYING \
 	COPYING-LGPL-2.1 \
