--- mozilla/xpfe/components/prefwindow/resources/content/preftree.xul.debug	Fri Jan  4 14:48:46 2002
+++ mozilla/xpfe/components/prefwindow/resources/content/preftree.xul	Fri Jan  4 14:49:18 2002
@@ -151,24 +151,6 @@
         </outlinerchildren>
       </outlineritem>
 
-      <outlineritem container="true" open="true" id="debugItem">
-        <outlinerrow>       
-          <outlinercell url="chrome://communicator/content/pref/pref-debug.xul" label="&debug.label;"/> 
-        </outlinerrow>
-        <outlinerchildren id="debugChildren">
-          <outlineritem>
-            <outlinerrow>
-              <outlinercell url="chrome://communicator/content/pref/pref-debug1.xul" label="&debug1.label;"/> 
-            </outlinerrow>
-          </outlineritem>
-          <outlineritem>
-            <outlinerrow>
-              <outlinercell url="chrome://communicator/content/pref/pref-debug2.xul" label="&debug2.label;"/> 
-            </outlinerrow>
-          </outlineritem>
-        </outlinerchildren>
-      </outlineritem>
-
       <outlineritem id="offlineItem">
         <outlinerrow>
           <outlinercell url="chrome://communicator/content/pref/pref-offline.xul" label="&offline.label;"/> 
