--- mozilla/editor/ui/composer/content/editor.xul.debug	Sun Aug  5 15:49:15 2001
+++ mozilla/editor/ui/composer/content/editor.xul	Sun Aug  5 15:49:32 2001
@@ -155,7 +155,7 @@
     <menu id="tasksMenu" accesskey="t"/>
 
     <!-- DEBUG only -->
-    <menu id="debugMenu" /> 
+    <!-- <menu id="debugMenu" /> -->
     <!-- end DEBUG only -->
   
     <!-- help menu filled from globalOverlay -->
--- mozilla/editor/ui/composer/content/editorOverlay.xul.debug	Sun Aug  5 15:47:09 2001
+++ mozilla/editor/ui/composer/content/editorOverlay.xul	Sun Aug  5 15:48:58 2001
@@ -804,6 +804,7 @@
   <menuitem id="menu_pasteQuote" accesskey="&editpastequotation.accesskey;" observes="cmd_pasteQuote"/>
 
   <!-- DEBUG only -->
+  <!--
   <menu id="debugMenu" label="&debugMenu.label;">
   <menupopup>
     <menuitem label="&newEditorTestPage.label;" oncommand="window.openDialog('chrome://editor/content','_blank','chrome,all,dialog=no','chrome://editor/content/EditorInitPage.html')"/>
@@ -857,6 +858,7 @@
             oncommand="window.focus()"/>
   </menupopup>
   </menu>
+  -->
   <!-- end DEBUG only -->
 
 </overlay>
--- mozilla/editor/ui/composer/content/TextEditorAppShell.xul.debug	Sun Aug  5 15:49:45 2001
+++ mozilla/editor/ui/composer/content/TextEditorAppShell.xul	Sun Aug  5 15:49:55 2001
@@ -113,7 +113,7 @@
       <menu accesskey="t" id="tasksMenu"/>
 
       <!-- DEBUG only -->
-      <menu id="debugMenu" />
+      <!-- <menu id="debugMenu" /> -->
       <!-- end DEBUG only -->
 
       <!-- help menu filled from globalOverlay -->
