--- mozilla/xpfe/components/prefwindow/resources/content/preftree.xul.debug	Wed Jan 16 13:15:13 2002
+++ mozilla/xpfe/components/prefwindow/resources/content/preftree.xul	Wed Jan 16 13:15:29 2002
@@ -151,24 +151,6 @@
       </outlinerchildren>
     </outlineritem>
 
-    <outlineritem container="true" id="debugItem">
-      <outlinerrow>       
-        <outlinercell url="chrome://communicator/content/pref/pref-debug.xul" label="&debug.label;"/> 
-      </outlinerrow>
-      <outlinerchildren id="debugChildren">
-        <outlineritem>
-          <outlinerrow>
-            <outlinercell url="chrome://communicator/content/pref/pref-debug1.xul" label="&debug1.label;"/> 
-          </outlinerrow>
-        </outlineritem>
-        <outlineritem>
-          <outlinerrow>
-            <outlinercell url="chrome://communicator/content/pref/pref-debug2.xul" label="&debug2.label;"/> 
-          </outlinerrow>
-        </outlineritem>
-      </outlinerchildren>
-    </outlineritem>
-
     <outlineritem id="offlineItem">
       <outlinerrow>
         <outlinercell url="chrome://communicator/content/pref/pref-offline.xul" label="&offline.label;"/> 
