--- mozilla/xpfe/communicator/resources/locale/en-US/taskbar.rdf.non-mozilla	Mon Jan 15 14:41:41 2001
+++ mozilla/xpfe/communicator/resources/locale/en-US/taskbar.rdf	Mon Jan 15 14:42:10 2001
@@ -1,47 +1,12 @@
 <?xml version="1.0"?>
 <!DOCTYPE RDF 
 [
-<!-- LOCALIZATION NOTE  (mozillaButton.label): DONT_TRANSLATE -->
-<!ENTITY  mozillaButton.label "Mozilla">
-<!ENTITY  mozillaTooltip.label  "Links to mozilla.org webtools">
-<!-- taskbar popups -->
-<!-- LOCALIZATION NOTE  (bugzillaButton.label): Do NOT translate "Bugzilla" -->
-<!ENTITY  bugzillaButton.label "Go to Bugzilla">
-<!ENTITY  tinderboxButton.label "Look at Tinderbox">
-<!-- LOCALIZATION NOTE  (bonsai.label): Do NOT translate "Bonsai" -->
-<!ENTITY  bonsai.label "Look at Bonsai">
-
 ]
 >
 <RDF:RDF xmlns:RDF="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
          xmlns:NC="http://home.netscape.com/NC-rdf#">
 
 <!-- root start  -->
-  <RDF:Seq ID="root">
-
-    <RDF:li>
-      <RDF:Seq about="urn:taskbar:mozilla-group">
-        <RDF:li resource="urn:taskbar:mozilla-group:bonsai" />
-        <RDF:li resource="urn:taskbar:mozilla-group:bugzilla" />
-     </RDF:Seq>
-    </RDF:li>
-
-  </RDF:Seq>
 <!-- root end -->
 
-<!-- Description -->
-<RDF:Description about="urn:taskbar:mozilla-group">
-    <NC:title>&mozillaButton.label;</NC:title>
-    <NC:tooltip>&mozillaTooltip.label;</NC:tooltip>
-</RDF:Description>
-
-  <RDF:Description about="urn:taskbar:mozilla-group:bonsai">
-    <NC:title>&bonsai.label;</NC:title> 
-    <NC:content>openTopWin('http://bonsai.mozilla.org/toplevel.cgi?treeid=SeaMonkey')</NC:content>
-  </RDF:Description>
-
-  <RDF:Description about="urn:taskbar:mozilla-group:bugzilla">
-    <NC:title>&bugzillaButton.label;</NC:title> 
-    <NC:content>openTopWin('http://bugzilla.mozilla.org')</NC:content>
-  </RDF:Description>
 </RDF:RDF>
