diff --git a/browser/chrome/browser/content/browser/aboutDialog.xhtml b/browser/chrome/browser/content/browser/aboutDialog.xhtml
index 0914b7f..33a3d74 100644
--- ./browser/chrome/browser/content/browser/aboutDialog.xhtml
+++ ./browser/chrome/browser/content/browser/aboutDialog.xhtml
@@ -25,6 +25,9 @@
 
     <html:link rel="localization" href="branding/brand.ftl"/>
     <html:link rel="localization" href="browser/aboutDialog.ftl"/>
+    <!-- @nora:inject:start -->
+    <html:link rel="localization" href="browser/floorp.ftl"/>
+    <!-- @nora:inject:end -->
   </linkset>
 
   <html:div id="aboutDialogContainer">
@@ -94,26 +97,23 @@
               </description>
             </vbox>
           </hbox>
-
-          <description class="text-blurb" id="currentChannelText"
-                       data-l10n-id="aboutdialog-channel-description"
-                       data-l10n-attrs="{&quot;channel&quot;: &quot;&quot;}">
-            <label id="currentChannel" data-l10n-name="current-channel"/>
+          <!-- @nora:inject:start -->
+          <description class="text-blurb" id="communityDesc" data-l10n-id="about-floorp">
+            <label is="text-link" href="https://floorp.app" data-l10n-name="floorp-browser-link"/>
+            <label is="text-link" href="https://ablaze.one" data-l10n-name="ablaze-Link"/>
+            <label is="text-link" href="https://github.com/sponsors/Ablaze-MIRAI" data-l10n-name="helpus-donateLink"/>
+          <!-- @nora:inject:end -->
           </description>
-          <vbox id="experimental" hidden="true">
-            <description class="text-blurb" id="warningDesc" data-l10n-id="warningDesc-version"></description>
-            <description class="text-blurb" id="communityExperimentalDesc" data-l10n-id="community-exp">
-              <label is="text-link" href="https://www.mozilla.org/?utm_source=firefox-browser&#38;utm_medium=firefox-desktop&#38;utm_campaign=about-dialog" data-l10n-name="community-exp-mozillaLink"/>
-              <label is="text-link" useoriginprincipal="true" href="about:credits" data-l10n-name="community-exp-creditsLink"/>
-            </description>
-          </vbox>
-          <description class="text-blurb" id="communityDesc" data-l10n-id="community-2">
-            <label is="text-link" href="https://www.mozilla.org/?utm_source=firefox-browser&#38;utm_medium=firefox-desktop&#38;utm_campaign=about-dialog" data-l10n-name="community-mozillaLink"/>
-            <label is="text-link" useoriginprincipal="true" href="about:credits" data-l10n-name="community-creditsLink"/>
+          <!-- @nora:inject:start -->
+          <description class="text-blurb" id="floorp-contributor" data-l10n-id="icon-creator">
+            <label is="text-link" href="https://twitter.com/mwxdxx" data-l10n-name="brand-logo-twitter"/>
+            <label is="text-link" href="https://twitter.com/CutterKnife_" data-l10n-name="browser-logo-twitter"/>
+          <!-- @nora:inject:end -->
           </description>
-          <description class="text-blurb" id="contributeDesc" data-l10n-id="helpus">
-            <label is="text-link" href="https://foundation.mozilla.org/?form=firefox-about" data-l10n-name="helpus-donateLink"/>
-            <label is="text-link" href="https://www.mozilla.org/contribute/?utm_source=firefox-browser&#38;utm_medium=firefox-desktop&#38;utm_campaign=about-dialog" data-l10n-name="helpus-getInvolvedLink"/>
+          <!-- @nora:inject:start -->
+          <description class="text-blurb" id="floorp-contributor" data-l10n-id="contributors">
+            <label is="text-link" useoriginprincipal="true" href="about:contributors" data-l10n-name="about-contributor"/>
+          <!-- @nora:inject:end -->
           </description>
         </vbox>
       </vbox>
@@ -121,8 +121,10 @@
     <vbox id="bottomBox">
       <hbox pack="center">
         <label is="text-link" class="bottom-link" useoriginprincipal="true" href="about:license" data-l10n-id="bottomLinks-license"/>
-        <label is="text-link" class="bottom-link" useoriginprincipal="true" href="about:rights" data-l10n-id="bottomLinks-rights"/>
-        <label is="text-link" class="bottom-link" href="https://www.mozilla.org/privacy/?utm_source=firefox-browser&#38;utm_medium=firefox-desktop&#38;utm_campaign=about-dialog" data-l10n-id="bottomLinks-privacy"/>
+        <!-- @nora:inject:start -->
+        <label is="text-link" class="bottom-link" useoriginprincipal="true" href="https://github.com/Floorp-Projects/About-Floorp-Projects/blob/main/UserLicense.md" data-l10n-id="bottomLinks-rights"/>
+        <label is="text-link" class="bottom-link" href="https://docs.ablaze.one/floorp_privacy_policy/" data-l10n-id="bottomLinks-privacy"/>
+        <!-- @nora:inject:end -->
       </hbox>
       <description id="trademark" data-l10n-id="trademarkInfo"></description>
     </vbox>
