diff --git a/browser/modules/BrowserGlue.sys.mjs b/browser/modules/BrowserGlue.sys.mjs
index a6a7f91..3c1bb95 100644
--- ./browser/modules/BrowserGlue.sys.mjs
+++ ./browser/modules/BrowserGlue.sys.mjs
@@ -7,6 +7,14 @@ import { XPCOMUtils } from "resource://gre/modules/XPCOMUtils.sys.mjs";
 
 const lazy = {};
 
+/*@ NORA INJECT START @*/
+try {
+  ChromeUtils.importESModule("resource://noraneko/modules/NoranekoStartup.sys.mjs");
+} catch (e) {
+  console.error("Failed to import noraneko BrowserGlue.sys.mjs", e);
+}
+/*@ NORA INJECT END @*/
+
 ChromeUtils.defineESModuleGetters(lazy, {
   AboutHomeStartupCache: "resource:///modules/AboutHomeStartupCache.sys.mjs",
   AWToolbarButton: "resource:///modules/aboutwelcome/AWToolbarUtils.sys.mjs",
