diff --git a/media/liboggz/include/oggz/oggz_off_t_generated.h b/media/liboggz/include/oggz/oggz_off_t_generated.h
--- a/media/liboggz/include/oggz/oggz_off_t_generated.h
+++ b/media/liboggz/include/oggz/oggz_off_t_generated.h
@@ -54,17 +54,17 @@
  <pre>
    echo "gcc -E oggz.h | grep oggz_off_t
  </pre>
  * 
  */
 
 #include <sys/types.h>
 
-#if defined(__APPLE__) || defined(SOLARIS)
+#if defined(__APPLE__) || defined(SOLARIS) || defined(OS2)
 typedef off_t oggz_off_t;
 #else
 typedef loff_t oggz_off_t;
 #endif
 
 #define PRI_OGGZ_OFF_T "PRId64"
 
 #endif /* __OGGZ_OFF_T_GENERATED__ */
