diff --git a/media/liboggz/src/liboggz/oggz_seek.c b/media/liboggz/src/liboggz/oggz_seek.c
--- a/media/liboggz/src/liboggz/oggz_seek.c
+++ b/media/liboggz/src/liboggz/oggz_seek.c
@@ -717,17 +717,17 @@ oggz_seek_set (OGGZ * oggz, ogg_int64_t 
 					      &serialno);
       unit_end = oggz_get_unit (oggz, serialno, granule_at);
 #ifdef DEBUG
       printf ("oggz_seek_set: [C] offset_next @%" PRI_OGGZ_OFF_T "d, g%lld, (s%ld)\n",
 	      offset_next, granule_at, serialno);
       printf ("oggz_seek_set: [c] u%lld\n",
 	      oggz_get_unit (oggz, serialno, granule_at));
 #endif
-    } else {
+    } else if (offset_next >= 0) {
       serialno = ogg_page_serialno (og);
       granule_at = ogg_page_granulepos (og);
     }
 
     if (offset_next < 0) {
       goto notfound;
     }
 
