diff --git a/ipc/chromium/src/third_party/libevent/android/event2/event-config.h b/ipc/chromium/src/third_party/libevent/android/event2/event-config.h
--- a/ipc/chromium/src/third_party/libevent/android/event2/event-config.h
+++ b/ipc/chromium/src/third_party/libevent/android/event2/event-config.h
@@ -19,23 +19,31 @@
 
 /* Define if libevent should not allow replacing the mm functions */
 /* #undef EVENT__DISABLE_MM_REPLACEMENT */
 
 /* Define if libevent should not be compiled with thread support */
 /* #undef EVENT__DISABLE_THREAD_SUPPORT */
 
 /* Define to 1 if you have the `accept4' function. */
+#if __ANDROID_API__ >= 21
 #define EVENT__HAVE_ACCEPT4 1
+#else
+/* #undef EVENT__HAVE_ACCEPT4 */
+#endif
 
 /* Define to 1 if you have the `arc4random' function. */
 #define EVENT__HAVE_ARC4RANDOM 1
 
 /* Define to 1 if you have the `arc4random_buf' function. */
+#if __ANDROID_API__ >= 21 || defined(__ANDROID_API_L__)
 #define EVENT__HAVE_ARC4RANDOM_BUF 1
+#else
+/* #undef EVENT__HAVE_ARC4RANDOM_BUF */
+#endif
 
 /* Define to 1 if you have the <arpa/inet.h> header file. */
 #define EVENT__HAVE_ARPA_INET_H 1
 
 /* Define to 1 if you have the `clock_gettime' function. */
 #define EVENT__HAVE_CLOCK_GETTIME 1
 
 /* Define to 1 if you have the declaration of `CTL_KERN', and to 0 if you
