diff --git a/ipc/chromium/src/third_party/libevent/signal.c b/ipc/chromium/src/third_party/libevent/signal.c
index 3f46295024e57..1eec621b48c6c 100644
--- a/ipc/chromium/src/third_party/libevent/signal.c
+++ b/ipc/chromium/src/third_party/libevent/signal.c
@@ -64,6 +64,8 @@
 #include "evmap-internal.h"
 #include "evthread-internal.h"
 
+#include "mozilla/Assertions.h"
+
 /*
   signal.c
 
@@ -280,6 +282,8 @@ evsig_add(struct event_base *base, evutil_socket_t evsignal, short old, short ev
 	struct evsig_info *sig = &base->sig;
 	(void)p;
 
+	MOZ_CRASH("Don't use this; see bug 1616462");
+
 	EVUTIL_ASSERT(evsignal >= 0 && evsignal < NSIG);
 
 	/* catch signals if they happen quickly */
