error[E0308]: mismatched types
  --> $DIR/type-mismatch.rs:36:9
   |
24 |     match &mut foo {
   |           -------- this expression has type `&mut type_mismatch::__EnumProjection<'_, {integer}, {integer}, _, _>`
...
36 |         None => {} //~ ERROR mismatched types
   |         ^^^^ expected enum `type_mismatch::__EnumProjection`, found enum `Option`
   |
   = note: expected enum `type_mismatch::__EnumProjection<'_, {integer}, {integer}, _, _>`
              found enum `Option<_>`

error[E0308]: mismatched types
  --> $DIR/type-mismatch.rs:68:9
   |
56 |     match &mut foo {
   |           -------- this expression has type `&mut type_mismatch_span_issue::__EnumProjection<'_, {integer}, {integer}, _, _>`
...
68 |         None => {} //~ ERROR mismatched types
   |         ^^^^ expected enum `type_mismatch_span_issue::__EnumProjection`, found enum `Option`
   |
   = note: expected enum `type_mismatch_span_issue::__EnumProjection<'_, {integer}, {integer}, _, _>`
              found enum `Option<_>`
