error: no rules expected the token `}`
 --> $DIR/unsupported.rs:4:21
  |
4 |     struct Struct1 {} //~ ERROR no rules expected the token `}`
  |                     ^ no rules expected this token in macro call

error: no rules expected the token `(`
 --> $DIR/unsupported.rs:8:19
  |
8 |     struct Struct2(); //~ ERROR no rules expected the token `(`
  |                   ^ no rules expected this token in macro call

error: no rules expected the token `;`
  --> $DIR/unsupported.rs:12:19
   |
12 |     struct Struct3; //~ ERROR no rules expected the token `;`
   |                   ^ no rules expected this token in macro call

error: no rules expected the token `enum`
  --> $DIR/unsupported.rs:16:5
   |
16 |     enum Enum { //~ ERROR no rules expected the token `enum`
   |     ^^^^ no rules expected this token in macro call

error: no rules expected the token `union`
  --> $DIR/unsupported.rs:22:5
   |
22 |     union Union { //~ ERROR no rules expected the token `union`
   |     ^^^^^ no rules expected this token in macro call
