Index: make/platform/clang_darwin.mk
===================================================================
--- make/platform/clang_darwin.mk	(revision 157958)
+++ make/platform/clang_darwin.mk	(working copy)
@@ -47,7 +47,7 @@
 # Configuration for targetting iOS, for some ARMv6 functions, which must be
 # in the same linkage unit, and for a couple of other functions that didn't
 # make it into libSystem.
-Configs += ios
+#Configs += ios
 UniversalArchs.ios := $(call CheckArches,i386 x86_64 armv6 armv7,ios)
 
 # Configuration for targetting OSX. These functions may not be in libSystem
@@ -56,13 +56,13 @@
 UniversalArchs.osx := $(call CheckArches,i386 x86_64,osx)
 
 # Configuration for use with kernel/kexts.
-Configs += cc_kext
+#Configs += cc_kext
 UniversalArchs.cc_kext := $(call CheckArches,armv6 armv7 i386 x86_64,cc_kext)
 
 # Configurations which define the profiling support functions.
 Configs += profile_osx
 UniversalArchs.profile_osx := $(call CheckArches,i386 x86_64,profile_osx)
-Configs += profile_ios
+#Configs += profile_ios
 UniversalArchs.profile_ios := $(call CheckArches,i386 x86_64 armv6 armv7,profile_ios)
 
 # Configurations which define the ASAN support functions.
