Index: lib/Driver/Tools.cpp
===================================================================
--- a/clang/lib/Driver/Tools.cpp	(revision 199443)
+++ b/clang/lib/Driver/Tools.cpp	(working copy)
@@ -1226,7 +1226,7 @@
   // All x86 devices running Android have core2 as their common
   // denominator. This makes a better choice than pentium4.
   if (Triple.getEnvironment() == llvm::Triple::Android)
-    return "core2";
+    return "686";
 
   // Everything else goes to x86-64 in 64-bit mode.
   if (Is64Bit)
