Index: lib/Driver/Tools.cpp
===================================================================
--- lib/Driver/Tools.cpp	(revision 167750)
+++ lib/Driver/Tools.cpp	(working copy)
@@ -1179,7 +1179,7 @@
       else if (getToolChain().getArch() == llvm::Triple::x86)
         // All x86 devices running Android have core2 as their common
         // denominator. This makes a better choice than pentium4.
-        CPUName = isAndroid ? "core2" : "pentium4";
+        CPUName = "i686";
     }
   }
 
