Index: lib/Driver/Tools.cpp
===================================================================
--- a/clang/lib/Driver/Tools.cpp        (revision 181515)
+++ b/clang/lib/Driver/Tools.cpp        (working copy)
@@ -1245,7 +1245,7 @@
   // All x86 devices running Android have core2 as their common
   // denominator. This makes a better choice than pentium4.
   if (Triple.getEnvironment() == llvm::Triple::Android)
-    return "core2";
+    return "686";

   // Fallback to p4.
   return "pentium4";
