diff -up4 en-US.dic en-US.dic
--- en-US.dic	Sat Nov 24 21:50:56 2007
+++ en-US.dic	Sat Dec  8 10:05:34 2007
@@ -2300,8 +2300,9 @@ Budweiser/M
 Buffalo/M
 Buffy/M
 Buford/M
 Bugatti/M
+Bugzilla/M
 Buick/M
 Buiron/M
 Bujumbura/M
 Bukhara
@@ -2503,8 +2504,9 @@ Cami/M
 Camila/M
 Camile/M
 Camilla/M
 Camille/M
+Camino/M
 Cammi/M
 Cammie/M
 Cammy/M
 Camoens/M
@@ -2942,8 +2944,9 @@ Chateaubriand/M
 Chattahoochee/M
 Chattanooga/M
 Chatterley/M
 Chatterton
+ChatZilla/M
 Chaucer/M
 Chaunce/M
 Chauncey/M
 Chautauqua
@@ -6325,8 +6328,9 @@ HS
 HST
 HT
 HTML/M
 HTTP
+HTTPS
 HUD/M
 Ha
 Haas/M
 Habakkuk
@@ -6945,8 +6949,9 @@ Hun/SM
 Hunfredo/M
 Hung
 Hungarian/SM
 Hungary/M
+Hunspell/M
 Hunt/R
 Hunter/M
 Huntington/M
 Huntlee/M
@@ -9504,8 +9509,9 @@ Mandela
 Mandelbrot
 Mandi/M
 Mandie/M
 Mandingo
+Mandriva/M
 Mandy/M
 Manet
 Manfred/M
 Manhattan/SM
@@ -10619,8 +10625,9 @@ Myrtie/M
 Myrtle/M
 Myrvyn/M
 Myrwyn/M
 Mysore
+MySpell/M
 Myst/M
 N'Djamena
 N/MD
 NAACP
@@ -13240,8 +13247,9 @@ Scythia
 Scythian
 Se/MH
 Seaborg
 Seagram/M
+SeaMonkey/M
 Seamus/M
 Sean/M
 Seana/M
 Sears/M
@@ -13979,8 +13987,9 @@ Sumner/M
 Sumter
 Sun/SM
 Sunbeam/M
 Sunbelt/M
+Sunbird/M
 Sundanese/M
 Sundas
 Sunday/MS
 Sung
@@ -42230,8 +42239,10 @@ philosophical/Y
 philosophize/ZGDRS
 philosophizer/M
 philosophy/SM
 philter/MS
+phish/DGS
+phisher/MS
 phlebitis/M
 phlebotomy
 phlegm/M
 phlegmatic
@@ -49102,8 +49113,9 @@ spaghetti/M
 spake
 spam/S
 spamblock/S
 spammed
+spammer/MS
 spamming
 span/MS
 spandex/M
 spandrels
@@ -53649,8 +53661,10 @@ unilateralist
 unimportance
 unimportant
 unimpressive
 uninhibited/Y
+uninstall/GSBD
+uninstaller/MS
 uninsured
 unintellectual
 unintelligent
 unintended
@@ -53811,8 +53825,10 @@ unstamped
 unsteady/PTR
 unstinting/Y
 unstoppably
 unstrapping
+unsubscribe/DGS
+unsubscriber/MS
 unsubstantial
 unsubtle
 unsure/P
 unsuspecting/Y
@@ -54984,9 +55000,10 @@ web/SM
 webbed
 webbing/M
 webfeet
 webfoot/M
-webmaster/S
+webmaster/MS
+webmistress/S
 website/SM
 wed/AS
 wedded/A
 wedder
