--- en-US.aff	Sat Nov 24 21:50:56 2007
+++ en-US.aff	Sat Nov 29 11:02:17 2008
@@ -109,15 +109,19 @@ SFX B   e     able       [^aeiou]e
 
 SFX L Y 1
 SFX L   0     ment       .
 
-REP 88
+SFX i N 1
+SFX i   us    i          us
+
+REP 90
 REP a ei
 REP ei a
 REP a ey
 REP ey a
 REP ai ie
 REP ie ai
+REP alot a_lot
 REP are air
 REP are ear
 REP are eir
 REP air are
@@ -198,4 +202,5 @@ REP z ss
 REP ss z
 REP shun tion
 REP shun sion
 REP shun cion
+REP sitted sat
--- en-US.dic	Sat Nov 24 21:50:56 2007
+++ en-US.dic	Sat Nov 29 11:03:25 2008
@@ -1426,8 +1426,9 @@ Baotou/M
 Baptist/SM
 Baptiste/M
 Bar/H
 Barabbas
+Barack/M
 Barb/MR
 Barbabas/M
 Barbabra/M
 Barbadian/SM
@@ -1835,8 +1836,9 @@ Bible/MS
 Bic/M
 Biddie/M
 Biddle
 Biddy/M
+Biden/MS
 Bidget/M
 Bierce
 Bigfoot/M
 Biggles/M
@@ -2300,8 +2302,9 @@ Budweiser/M
 Buffalo/M
 Buffy/M
 Buford/M
 Bugatti/M
+Bugzilla/M
 Buick/M
 Buiron/M
 Bujumbura/M
 Bukhara
@@ -2503,8 +2506,9 @@ Cami/M
 Camila/M
 Camile/M
 Camilla/M
 Camille/M
+Camino/M
 Cammi/M
 Cammie/M
 Cammy/M
 Camoens/M
@@ -2942,8 +2946,9 @@ Chateaubriand/M
 Chattahoochee/M
 Chattanooga/M
 Chatterley/M
 Chatterton
+ChatZilla/M
 Chaucer/M
 Chaunce/M
 Chauncey/M
 Chautauqua
@@ -6325,8 +6330,9 @@ HS
 HST
 HT
 HTML/M
 HTTP
+HTTPS
 HUD/M
 Ha
 Haas/M
 Habakkuk
@@ -6418,9 +6424,9 @@ Haney/M
 Hangul/M
 Hangzhou/M
 Hank/M
 Hanna/M
-Hannah
+Hannah/M
 Hanni/SM
 Hannibal/M
 Hannie/M
 Hanny/M
@@ -6945,8 +6951,9 @@ Hun/SM
 Hunfredo/M
 Hung
 Hungarian/SM
 Hungary/M
+Hunspell/M
 Hunt/R
 Hunter/M
 Huntington/M
 Huntlee/M
@@ -7358,9 +7365,9 @@ Jamaican/SM
 Jamal/M
 Jamar/M
 Jame/SM
 Jamel/M
-James/M
+James
 Jameson
 Jamestown
 Jamesy/M
 Jamey/M
@@ -7377,9 +7384,9 @@ Jana/M
 Janacek/M
 Janaya/M
 Janaye/M
 Jandy/M
-Jane
+Jane/M
 Janean/M
 Janeczka/M
 Janeen/M
 Janek/M
@@ -7390,9 +7397,9 @@ Janella/M
 Janelle/M
 Janene/M
 Janenna/M
 Janessa/M
-Janet
+Janet/M
 Janeta/M
 Janetta/M
 Janette/M
 Janeva/M
@@ -9504,8 +9511,9 @@ Mandela
 Mandelbrot
 Mandi/M
 Mandie/M
 Mandingo
+Mandriva/M
 Mandy/M
 Manet
 Manfred/M
 Manhattan/SM
@@ -10619,8 +10627,9 @@ Myrtie/M
 Myrtle/M
 Myrvyn/M
 Myrwyn/M
 Mysore
+MySpell/M
 Myst/M
 N'Djamena
 N/MD
 NAACP
@@ -11159,8 +11168,9 @@ Oates
 Oaxaca/M
 Ob/MD
 Obadiah
 Obadias/M
+Obama/MS
 Obed/M
 Obediah/M
 Oberlin/M
 Oberon
@@ -11501,8 +11511,9 @@ Palestine/M
 Palestinian/SM
 Palestrina
 Paley
 Palikir/M
+Palin/MS
 Palisades/M
 Pall/M
 Palladio
 Palm/MR
@@ -13240,8 +13251,9 @@ Scythia
 Scythian
 Se/MH
 Seaborg
 Seagram/M
+SeaMonkey/M
 Seamus/M
 Sean/M
 Seana/M
 Sears/M
@@ -13979,8 +13991,9 @@ Sumner/M
 Sumter
 Sun/SM
 Sunbeam/M
 Sunbelt/M
+Sunbird/M
 Sundanese/M
 Sundas
 Sunday/MS
 Sung
@@ -14424,9 +14437,9 @@ Thieu/M
 Thimbu
 Thimphu
 Thom/M
 Thoma/SM
-Thomas/M
+Thomas
 Thomasa/M
 Thomasin/M
 Thomasina/M
 Thomasine/M
@@ -14895,9 +14908,8 @@ Unitarian/MS
 Unitarianism/MS
 Unitas/M
 Unix/S
 Unukalhai/M
-Ununtu/M
 Upanishads
 Updike
 Upjohn/M
 Upton/M
@@ -15437,8 +15449,9 @@ Wiemar/M
 Wiesel/M
 Wiesenthal/M
 Wiggins
 Wigner/M
+Wikipedia/M
 Wilberforce
 Wilbert/M
 Wilbur/M
 Wilburn/M
@@ -17628,8 +17641,9 @@ arcana
 arcane/PY
 arch/PZTGVMDRSY
 archaeological/Y
 archaeologist/SM
+archaeology/M
 archaic
 archaically
 archaism/MS
 archaist/MS
@@ -17642,8 +17656,10 @@ archdiocesan
 archdiocese/SM
 archduchess/MS
 archduke/MS
 archenemy/SM
+archeological/Y
+archeologist/SM
 archeology/M
 archeopteryx
 archer/M
 archery/M
@@ -18309,8 +18325,9 @@ awning/M
 awoke
 awoken
 awry
 ax/MDSG
+axe/M
 axehead/S
 axeman
 axial/Y
 axillary
@@ -20700,10 +20717,9 @@ cachet/MS
 cackle/MZGDRS
 cackler/M
 cacophonous
 cacophony/SM
-cacti
-cactus/M
+cactus/Mi
 cad/SM
 cadaver/SM
 cadaverous
 caddish/YP
@@ -25527,9 +25543,10 @@ dialectic/SM
 dialectical/Y
 dialectics/M
 dialer
 dialing/S
-dialog/SM
+dialog/SMGD
+dialogue/SMRGD
 dialysis/M
 dialyzes
 diam
 diamante
@@ -27384,8 +27401,9 @@ encumber/EGSD
 encumbered/U
 encumbrance/SM
 ency
 encyclical/SM
+encyclopaedia
 encyclopedia/MS
 encyclopedic
 encyst/LSGD
 encystment/M
@@ -29493,11 +29511,10 @@ foamy/RTP
 fob/SM
 fobbed
 fobbing
 focal/Y
-focus's
-focus/ADSG
-focused/U
+focus/ACGRSMBi
+focused/UC
 fodder/SM
 foe/SM
 foetid
 fog's
@@ -36323,9 +36340,8 @@ locatable/A
 locate/EAGNVDS
 location's/A
 location/ESM
 locational
-loci
 lock/MDRSBZG
 locked/A
 locker/M
 locket/MS
@@ -36339,9 +36355,9 @@ loco/S
 locomotion/M
 locomotive/MS
 locoweed/SM
 locum/S
-locus/M
+locus/Mi
 locust/SM
 locution/MS
 lode/MS
 lodestar/MS
@@ -37405,8 +37421,9 @@ megalomania/M
 megalomaniac/SM
 megalopolis/MS
 megaparsec
 megaphone/DSMG
+megapixel
 megastar/S
 megaton/SM
 megawatt/MS
 meiosis/M
@@ -40140,9 +40157,9 @@ octant
 octave/MS
 octavo/MS
 octet/SM
 octogenarian/MS
-octopus/MS
+octopus/MSi
 ocular/MS
 oculist/SM
 odalisque/SM
 odd/STRYLP
@@ -42230,8 +42247,10 @@ philosophical/Y
 philosophize/ZGDRS
 philosophizer/M
 philosophy/SM
 philter/MS
+phish/DGS
+phisher/MS
 phlebitis/M
 phlebotomy
 phlegm/M
 phlegmatic
@@ -43400,10 +43419,11 @@ predominant/Y
 predominate/DSYG
 preemie/SM
 preeminence/M
 preeminent/Y
-preempt/GVSD
+preempt/GSD
 preemption/M
+preemptive/Y
 preen/DSG
 preexist/DGS
 preexistence/M
 pref
@@ -44994,8 +45014,9 @@ receivables/M
 receive/DRSZGB
 received/U
 receiver/M
 receivership/M
+recency
 recension
 recent/YTP
 recentness/M
 receptacle/MS
@@ -49102,8 +49123,9 @@ spaghetti/M
 spake
 spam/S
 spamblock/S
 spammed
+spammer/MS
 spamming
 span/MS
 spandex/M
 spandrels
@@ -53649,8 +53671,10 @@ unilateralist
 unimportance
 unimportant
 unimpressive
 uninhibited/Y
+uninstall/GSBD
+uninstaller/MS
 uninsured
 unintellectual
 unintelligent
 unintended
@@ -53811,8 +53835,10 @@ unstamped
 unsteady/PTR
 unstinting/Y
 unstoppably
 unstrapping
+unsubscribe/DGS
+unsubscriber/MS
 unsubstantial
 unsubtle
 unsure/P
 unsuspecting/Y
@@ -54984,9 +55010,10 @@ web/SM
 webbed
 webbing/M
 webfeet
 webfoot/M
-webmaster/S
+webmaster/MS
+webmistress/S
 website/SM
 wed/AS
 wedded/A
 wedder
@@ -55310,8 +55337,9 @@ wiglet/SM
 wigwag/SM
 wigwagged
 wigwagging
 wigwam/SM
+wiki/SM
 wild/MRYSTP
 wildcat/MS
 wildcatted
 wildcatter/MS
