diff --git a/cmd/shlibsign/Makefile b/cmd/shlibsign/Makefile
--- a/cmd/shlibsign/Makefile
+++ b/cmd/shlibsign/Makefile
@@ -40,19 +40,19 @@ EXTRA_SHARED_LIBS += \
 	-lplds4 \
 	-lnspr4 \
 	$(NULL)
 
 endif
 
 
 # sign any and all shared libraries that contain the word freebl
-ifeq ($(NSS_BUILD_WITHOUT_SOFTOKEN),1))
-CHECKLIBS=
-CHECKLOC=
+ifeq ($(NSS_BUILD_WITHOUT_SOFTOKEN),1)
+CHECKLIBS =
+CHECKLOC =
 else
 CHECKLIBS = $(DIST)/lib/$(DLL_PREFIX)softokn3.$(DLL_SUFFIX)
 CHECKLIBS += $(wildcard $(DIST)/lib/$(DLL_PREFIX)freebl*3.$(DLL_SUFFIX))
 ifndef NSS_DISABLE_DBM
 CHECKLIBS += $(DIST)/lib/$(DLL_PREFIX)nssdbm3.$(DLL_SUFFIX)
 endif
 CHECKLOC = $(CHECKLIBS:.$(DLL_SUFFIX)=.chk)
 
