diff -up security/nss/cmd/platlibs.mk.pre519550 security/nss/cmd/platlibs.mk
--- security/nss/cmd/platlibs.mk.pre519550	2010-02-06 16:46:29.118236532 +0100
+++ security/nss/cmd/platlibs.mk	2010-02-06 16:38:20.588959438 +0100
@@ -110,6 +110,7 @@ PKIXLIB = \
 
 # can't do this in manifest.mn because OS_ARCH isn't defined there.
 ifeq (,$(filter-out WINNT WINCE,$(OS_ARCH))) 
+SQLITE = $(LIB_PREFIX)sqlite3.$(LIB_SUFFIX)
 
 EXTRA_LIBS += \
 	$(DIST)/lib/$(LIB_PREFIX)smime.$(LIB_SUFFIX) \
@@ -130,7 +131,7 @@ EXTRA_LIBS += \
 	$(DIST)/lib/$(LIB_PREFIX)nssb.$(LIB_SUFFIX) \
 	$(PKIXLIB) \
 	$(DBMLIB) \
-	$(DIST)/lib/$(LIB_PREFIX)sqlite3.$(LIB_SUFFIX) \
+	$(DIST)/lib/$(SQLITE) \
 	$(DIST)/lib/$(LIB_PREFIX)nssutil3.$(LIB_SUFFIX) \
 	$(NSPR_LIB_DIR)/$(NSPR31_LIB_PREFIX)plc4.$(LIB_SUFFIX) \
 	$(NSPR_LIB_DIR)/$(NSPR31_LIB_PREFIX)plds4.$(LIB_SUFFIX) \
diff -up security/nss/lib/softoken/config.mk.pre519550 security/nss/lib/softoken/config.mk
--- security/nss/lib/softoken/config.mk.pre519550	2010-02-06 16:46:44.701986018 +0100
+++ security/nss/lib/softoken/config.mk	2010-02-06 16:51:10.634999945 +0100
@@ -53,11 +53,12 @@ RES = $(OBJDIR)/$(LIBRARY_NAME).res
 RESNAME = $(LIBRARY_NAME).rc
 
 ifdef NS_USE_GCC
+SQLITE = -lsqlite3
 EXTRA_SHARED_LIBS += \
 	-L$(DIST)/lib \
 	-L$(NSSUTIL_LIB_DIR) \
 	-lnssutil3 \
-	-lsqlite3 \
+	$(SQLITE) \
 	-L$(NSPR_LIB_DIR) \
 	-lplc4 \
 	-lplds4 \
@@ -65,8 +66,9 @@ EXTRA_SHARED_LIBS += \
 	$(NULL)
 else # ! NS_USE_GCC
 
+SQLITE = sqlite3.lib
 EXTRA_SHARED_LIBS += \
-	$(DIST)/lib/sqlite3.lib \
+	$(DIST)/lib/$(SQLITE) \
 	$(DIST)/lib/nssutil3.lib \
 	$(NSPR_LIB_DIR)/$(NSPR31_LIB_PREFIX)plc4.lib \
 	$(NSPR_LIB_DIR)/$(NSPR31_LIB_PREFIX)plds4.lib \
@@ -79,11 +81,12 @@ else
 
 # $(PROGRAM) has NO explicit dependencies on $(EXTRA_SHARED_LIBS)
 # $(EXTRA_SHARED_LIBS) come before $(OS_LIBS), except on AIX.
+SQLITE = -lsqlite3
 EXTRA_SHARED_LIBS += \
 	-L$(DIST)/lib \
 	-L$(NSSUTIL_LIB_DIR) \
 	-lnssutil3 \
-	-lsqlite3 \
+	$(SQLITE) \
 	-L$(NSPR_LIB_DIR) \
 	-lplc4 \
 	-lplds4 \
