Index: mozilla/security/coreconf/WIN32.mk
===================================================================
RCS file: /cvsroot/mozilla/security/coreconf/WIN32.mk,v
retrieving revision 1.39.2.1
diff -u -p -8 -r1.39.2.1 WIN32.mk
--- mozilla/security/coreconf/WIN32.mk	31 Jul 2010 04:23:37 -0000	1.39.2.1
+++ mozilla/security/coreconf/WIN32.mk	10 Aug 2010 20:46:44 -0000
@@ -126,17 +126,17 @@ else # !NS_USE_GCC
     ifdef BUILD_OPT
 	OS_CFLAGS  += -MD
 	ifeq (11,$(ALLOW_OPT_CODE_SIZE)$(OPT_CODE_SIZE))
 		OPTIMIZER += -O1
 	else
 		OPTIMIZER += -O2
 	endif
 	DEFINES    += -UDEBUG -U_DEBUG -DNDEBUG
-	DLLFLAGS   += -OUT:"$@"
+	DLLFLAGS   += -DYNAMICBASE -OUT:"$@"
 	ifdef MOZ_DEBUG_SYMBOLS
 		ifdef MOZ_DEBUG_FLAGS
 			OPTIMIZER += $(MOZ_DEBUG_FLAGS) -Fd$(OBJDIR)/
 		else
 			OPTIMIZER += -Zi -Fd$(OBJDIR)/
 		endif
 		DLLFLAGS += -DEBUG -OPT:REF
 		LDFLAGS += -DEBUG -OPT:REF
