Bug 1014976 - Don't make --enable-debug imply using the debug CRT in FFI

diff --git a/js/src/ctypes/libffi/msvcc.sh b/js/src/ctypes/libffi/msvcc.sh
--- a/js/src/ctypes/libffi/msvcc.sh
+++ b/js/src/ctypes/libffi/msvcc.sh
@@ -100,19 +100,23 @@ do
       shift 1
     ;;
     -g)
       # Enable debug symbol generation.
       args="$args -Zi"
       shift 1
     ;;
     -DFFI_DEBUG)
-      # Link against debug CRT and enable runtime error checks.
+      # Enable runtime error checks.
       args="$args -RTC1"
       defines="$defines $1"
+      shift 1
+    ;;
+    -DUSE_DEBUG_RTL)
+      # Link against debug CRT.
       md=-MDd
       shift 1
     ;;
     -c)
       args="$args -c"
       args="$(echo $args | sed 's%/Fe%/Fo%g')"
       single="-c"
       shift 1
