From: Michael Froman <mjfroman@mac.com>
Date: Thu, 26 May 2022 11:14:40 -0500
Subject: Bug 1766646 - (fix-3f418ccf99) disable a RTC_DCHECK_EQ for mozilla
 builds

---
 call/call.cc | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/call/call.cc b/call/call.cc
index a51e153064..86410b12de 100644
--- a/call/call.cc
+++ b/call/call.cc
@@ -1393,7 +1393,9 @@ void Call::DeliverRtcp(MediaType media_type, rtc::CopyOnWriteBuffer packet) {
   // and make sure that the flow of packets is consistent from the
   // `RtpTransport` class, via the *Channel and *Engine classes and into Call.
   // This way we'll also know more about the context of the packet.
+#if !defined(WEBRTC_MOZILLA_BUILD)
   RTC_DCHECK_EQ(media_type, MediaType::ANY);
+#endif
 
   // TODO(bugs.webrtc.org/11993): This should execute directly on the network
   // thread.
-- 
2.34.1

