diff --git a/media/libvpx/vpx/vp8.h b/media/libvpx/vpx/vp8.h
--- a/media/libvpx/vpx/vp8.h
+++ b/media/libvpx/vpx/vp8.h
@@ -41,33 +41,33 @@ enum vp8_com_control_id
     VP8_SET_REFERENCE           = 1,    /**< pass in an external frame into decoder to be used as reference frame */
     VP8_COPY_REFERENCE          = 2,    /**< get a copy of reference frame from the decoder */
     VP8_SET_POSTPROC            = 3,    /**< set the decoder's post processing settings  */
     VP8_SET_DBG_COLOR_REF_FRAME = 4,    /**< set the reference frames to color for each macroblock */
     VP8_SET_DBG_COLOR_MB_MODES  = 5,    /**< set which macro block modes to color */
     VP8_SET_DBG_COLOR_B_MODES   = 6,    /**< set which blocks modes to color */
     VP8_SET_DBG_DISPLAY_MV      = 7,    /**< set which motion vector modes to draw */
     VP8_COMMON_CTRL_ID_MAX,
-    VP8_DECODER_CTRL_ID_START   = 256,
+    VP8_DECODER_CTRL_ID_START   = 256
 };
 
 /*!\brief post process flags
  *
  * The set of macros define VP8 decoder post processing flags
  */
 enum vp8_postproc_level
 {
     VP8_NOFILTERING             = 0,
     VP8_DEBLOCK                 = 1<<0,
     VP8_DEMACROBLOCK            = 1<<1,
     VP8_ADDNOISE                = 1<<2,
     VP8_DEBUG_TXT_FRAME_INFO    = 1<<3, /**< print frame information */
     VP8_DEBUG_TXT_MBLK_MODES    = 1<<4, /**< print macro block modes over each macro block */
     VP8_DEBUG_TXT_DC_DIFF       = 1<<5, /**< print dc diff for each macro block */
-    VP8_DEBUG_TXT_RATE_INFO     = 1<<6, /**< print video rate info (encoder only) */
+    VP8_DEBUG_TXT_RATE_INFO     = 1<<6  /**< print video rate info (encoder only) */
 };
 
 /*!\brief post process flags
  *
  * This define a structure that describe the post processing settings. For
  * the best objective measure (using the PSNR metric) set post_proc_flag
  * to VP8_DEBLOCK and deblocking_level to 1.
  */
