diff --git a/media/libvpx/vp9/common/vp9_systemdependent.h b/media/libvpx/vp9/common/vp9_systemdependent.h
--- a/media/libvpx/vp9/common/vp9_systemdependent.h
+++ b/media/libvpx/vp9/common/vp9_systemdependent.h
@@ -12,17 +12,19 @@
 #define VP9_COMMON_VP9_SYSTEMDEPENDENT_H_
 
 #ifdef _MSC_VER
 # if _MSC_VER > 1310 && (defined(_M_X64) || defined(_M_IX86))
 #  include <intrin.h>
 #  define USE_MSC_INTRIN
 # endif
 # include <math.h>
-# define snprintf _snprintf
+# if _MSC_VER < 1900
+#  define snprintf _snprintf
+# endif
 #endif
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
 #include "./vpx_config.h"
 #if ARCH_X86 || ARCH_X86_64
