--- vpx/vpx_integer.h-	2014-09-17 15:49:58.000000000 -0700
+++ vpx/vpx_integer.h	2014-09-17 15:52:59.000000000 -0700
@@ -15,6 +15,8 @@
 /* get ptrdiff_t, size_t, wchar_t, NULL */
 #include <stddef.h>
 
+#if !defined(VPX_DONT_DEFINE_STDINT_TYPES)
+
 #if (defined(_MSC_VER) && (_MSC_VER < 1600)) || defined(VPX_EMULATE_INTTYPES)
 typedef signed char  int8_t;
 typedef signed short int16_t;
@@ -56,6 +58,8 @@
 
 #endif
 
+#endif // VPX_DONT_DEFINE_STDINT_TYPES
+
 /* VS2010 defines stdint.h, but not inttypes.h */
 #if defined(_MSC_VER) && _MSC_VER < 1800
 #define PRId64 "I64d"
