diff -r 5c557d4dd0c7 media/libvpx/vpx_ports/x86_abi_support.asm
--- a/media/libvpx/vpx_ports/x86_abi_support.asm Wed Jun 16 11:12:38 2010 +1200
+++ b/media/libvpx/vpx_ports/x86_abi_support.asm Thu Jun 17 15:09:49 2010 -0700
@@ -138,12 +138,16 @@
     %endmacro
   %endif
   %endif
+  %define HIDDEN_DATA
 %else
   %macro GET_GOT 1
   %endmacro
   %define GLOBAL wrt rip
   %ifidn __OUTPUT_FORMAT__,elf64
     %define WRT_PLT wrt ..plt
+    %define HIDDEN_DATA :data hidden
+  %else
+    %define HIDDEN_DATA
   %endif
 %endif
 %ifnmacro GET_GOT
diff -r 5c557d4dd0c7 media/libvpx/vp8/common/x86/subpixel_mmx.asm
--- a/media/libvpx/vp8/common/x86/subpixel_mmx.asm Wed Jun 16 11:12:38 2010 +1200
+++ b/media/libvpx/vp8/common/x86/subpixel_mmx.asm Thu Jun 17 15:09:49 2010 -0700
@@ -731,7 +731,7 @@
     times 4 dw 0x40
 
 align 16
-global sym(vp8_six_tap_mmx)
+global sym(vp8_six_tap_mmx) HIDDEN_DATA
 sym(vp8_six_tap_mmx):
     times 8 dw 0
     times 8 dw 0
@@ -791,7 +791,7 @@
 
 
 align 16
-global sym(vp8_bilinear_filters_mmx)
+global sym(vp8_bilinear_filters_mmx) HIDDEN_DATA
 sym(vp8_bilinear_filters_mmx):
     times 8 dw 128
     times 8 dw 0
