diff --git a/media/libvpx/vp8/common/x86/subpixel_sse2.asm b/media/libvpx/vp8/common/x86/subpixel_sse2.asm
--- a/media/libvpx/vp8/common/x86/subpixel_sse2.asm
+++ b/media/libvpx/vp8/common/x86/subpixel_sse2.asm
@@ -1003,17 +1003,17 @@ next_row8x8:
         paddw       xmm3,       xmm7
 
         movdqa      xmm7,       xmm4
 
         paddw       xmm3,       [rd GLOBAL]         ; xmm3 += round value
         psraw       xmm3,       VP8_FILTER_SHIFT        ; xmm3 /= 128
 
         packuswb    xmm3,       xmm0
-        movq        [rdi],      xmm3                 ; store the results in the destination
+        movq        QWORD PTR [rdi], xmm3           ; store the results in the destination
 
         add         rsp,        16                 ; next line
         add         rdi,        rdx
 
         cmp         rdi,        rcx
         jne         next_row8x8
 
     ;add rsp, 144
