diff --git a/media/libvpx/vp8/common/setupintrarecon.h b/media/libvpx/vp8/common/setupintrarecon.h
index e515c3a..9317a6d 100644
--- a/media/libvpx/vp8/common/setupintrarecon.h
+++ b/media/libvpx/vp8/common/setupintrarecon.h
@@ -3,16 +3,18 @@
  *
  *  Use of this source code is governed by a BSD-style license
  *  that can be found in the LICENSE file in the root of the source
  *  tree. An additional intellectual property rights grant can be found
  *  in the file PATENTS.  All contributing project authors may
  *  be found in the AUTHORS file in the root of the source tree.
  */
 
+#ifndef SETUPINTRARECON_H
+#define SETUPINTRARECON_H
 
 #include "vpx_scale/yv12config.h"
 extern void vp8_setup_intra_recon(YV12_BUFFER_CONFIG *ybf);
 extern void vp8_setup_intra_recon_top_line(YV12_BUFFER_CONFIG *ybf);
 
 static
 void setup_intra_recon_left(unsigned char *y_buffer,
                             unsigned char *u_buffer,
@@ -26,8 +28,10 @@ void setup_intra_recon_left(unsigned char *y_buffer,
         y_buffer[y_stride *i] = (unsigned char) 129;
 
     for (i = 0; i < 8; i++)
         u_buffer[uv_stride *i] = (unsigned char) 129;
 
     for (i = 0; i < 8; i++)
         v_buffer[uv_stride *i] = (unsigned char) 129;
 }
+
+#endif
diff --git a/media/libvpx/vpx_ports/vpx_once.h b/media/libvpx/vpx_ports/vpx_once.h
index 16a735c..0387a71 100644
--- a/media/libvpx/vpx_ports/vpx_once.h
+++ b/media/libvpx/vpx_ports/vpx_once.h
@@ -2,16 +2,19 @@
  *  Copyright (c) 2011 The WebM project authors. All Rights Reserved.
  *
  *  Use of this source code is governed by a BSD-style license
  *  that can be found in the LICENSE file in the root of the source
  *  tree. An additional intellectual property rights grant can be found
  *  in the file PATENTS.  All contributing project authors may
  *  be found in the AUTHORS file in the root of the source tree.
  */
+#ifndef VPX_ONCE_H
+#define VPX_ONCE_H
+
 #include "vpx_config.h"
 
 #if CONFIG_MULTITHREAD && defined(_WIN32)
 #include <windows.h>
 #include <stdlib.h>
 static void once(void (*func)(void))
 {
     static CRITICAL_SECTION *lock;
@@ -90,8 +93,10 @@ static void once(void (*func)(void))
 
     if(!done)
     {
         func();
         done = 1;
     }
 }
 #endif
+
+#endif
