--- va.h.old	2020-10-22 10:41:57.805112031 +0200
+++ va.h	2020-10-22 10:37:22.597088670 +0200
@@ -4613,18 +4613,18 @@
 #define VA_PICTURE_HEVC_RPS_LT_CURR             0x00000040
 
-#include <va/va_dec_hevc.h>
-#include <va/va_dec_jpeg.h>
+//#include <va/va_dec_hevc.h>
+//#include <va/va_dec_jpeg.h>
 #include "va_dec_vp8.h"
 #include "va_dec_vp9.h"
-#include <va/va_enc_hevc.h>
-#include <va/va_fei_hevc.h>
-#include <va/va_enc_h264.h>
-#include <va/va_enc_jpeg.h>
-#include <va/va_enc_mpeg2.h>
-#include <va/va_enc_vp8.h>
-#include <va/va_enc_vp9.h>
-#include <va/va_fei.h>
-#include <va/va_fei_h264.h>
-#include <va/va_vpp.h>
+//#include <va/va_enc_hevc.h>
+//#include <va/va_fei_hevc.h>
+//#include <va/va_enc_h264.h>
+//#include <va/va_enc_jpeg.h>
+//#include <va/va_enc_mpeg2.h>
+//#include <va/va_enc_vp8.h>
+//#include <va/va_enc_vp9.h>
+//#include <va/va_fei.h>
+//#include <va/va_fei_h264.h>
+//#include <va/va_vpp.h>
 
 /**@}*/
