diff --git a/nsprpub/pr/src/nspr.def b/nsprpub/pr/src/nspr.def
--- a/nsprpub/pr/src/nspr.def
+++ b/nsprpub/pr/src/nspr.def
@@ -457,8 +457,12 @@ EXPORTS ;-
 ;+} NSPR_4.9.2;
 ;+# Function PR_DuplicateEnvironment had been added in NSPR 4.10.9,
 ;+# but we neglected to add it to nspr.def until NSPR 4.12
 ;+NSPR_4.12 {
 ;+      global:
 		PR_DuplicateEnvironment;
 		PR_GetEnvSecure;
 ;+} NSPR_4.10.3;
+;+NSPR_4.17_fork {
+;+      global:
+		PR_EXPERIMENTAL_ONLY_IN_4_17_GetOverlappedIOHandle;
+;+} NSPR_4.16;
