diff --git a/llvm/utils/extract_symbols.py b/llvm/utils/extract_symbols.py
index 43f603963a2..01fe10d36f0 100755
--- a/llvm/utils/extract_symbols.py
+++ b/llvm/utils/extract_symbols.py
@@ -32,7 +32,7 @@ import argparse
 def dumpbin_get_symbols(lib):
     process = subprocess.Popen(['dumpbin','/symbols',lib], bufsize=1,
                                stdout=subprocess.PIPE, stdin=subprocess.PIPE,
-                               universal_newlines=True)
+                               universal_newlines=True, encoding='mbcs')
     process.stdin.close()
     for line in process.stdout:
         # Look for external symbols that are defined in some section
