--- a/extra/clang-tidy/tool/run-clang-tidy.py	2019-03-27 15:12:48.000000000 +0200
+++ b/extra/clang-tidy/tool/run-clang-tidy.py	2019-03-27 15:12:39.000000000 +0200
@@ -169,6 +169,7 @@
     with lock:
       sys.stdout.write(' '.join(invocation) + '\n' + output.decode('utf-8') + '\n')
       if len(err) > 0:
+        sys.stdout.flush()
         sys.stderr.write(err.decode('utf-8') + '\n')
     queue.task_done()
 
