diff --git a/compiler-rt/CMakeLists.txt b/compiler-rt/CMakeLists.txt
index 4f8e095c3b91..057dbf645450 100644
--- a/compiler-rt/CMakeLists.txt
+++ b/compiler-rt/CMakeLists.txt
@@ -69,6 +69,9 @@ set(COMPILER_RT_BAREMETAL_BUILD OFF CACHE BOOL
 
 if (COMPILER_RT_STANDALONE_BUILD)
   load_llvm_config()
+  set(CMAKE_CXX_STANDARD 17 CACHE STRING "C++ standard to conform to")
+  set(CMAKE_CXX_STANDARD_REQUIRED YES)
+  set(CMAKE_CXX_EXTENSIONS NO)
   if (TARGET intrinsics_gen)
     # Loading the llvm config causes this target to be imported so place it
     # under the appropriate folder in an IDE.
