Index: lib/CodeGen/AsmPrinter/AsmPrinter.cpp
===================================================================
--- a/llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp	(revision 226419)
+++ b/llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp	(working copy)
@@ -210,6 +210,8 @@
     OutStreamer->EmitFileDirective(M.getModuleIdentifier());
   }

+  OutStreamer->EmitCFISections(true, true);
+
   GCModuleInfo *MI = getAnalysisIfAvailable<GCModuleInfo>();
   assert(MI && "AsmPrinter didn't require GCModuleInfo?");
   for (auto &I : *MI)
