diff --git a/lld/MachO/LTO.cpp b/lld/MachO/LTO.cpp
index 2f5e9d06f396..7fbb41bef53d 100644
--- a/lld/MachO/LTO.cpp
+++ b/lld/MachO/LTO.cpp
@@ -65,8 +65,6 @@ static lto::Config createConfig() {
     pm.add(createObjCARCContractPass());
   };
 
-  c.AlwaysEmitRegularLTOObj = !config->ltoObjPath.empty();
-
   c.TimeTraceEnabled = config->timeTraceEnabled;
   c.TimeTraceGranularity = config->timeTraceGranularity;
   c.OptLevel = config->ltoo;
