diff --git a/media/libsydneyaudio/src/sydney_audio_alsa.c b/media/libsydneyaudio/src/sydney_audio_alsa.c
--- a/media/libsydneyaudio/src/sydney_audio_alsa.c
+++ b/media/libsydneyaudio/src/sydney_audio_alsa.c
@@ -138,17 +138,17 @@ sa_stream_open(sa_stream_t *s) {
                          SND_PCM_FORMAT_S16_LE,
 #else
                          SND_PCM_FORMAT_S16_BE,
 #endif
                          SND_PCM_ACCESS_RW_INTERLEAVED,
                          s->n_channels,
                          s->rate,
                          1,
-                         500000) < 0) {
+                         250000) < 0) {
     snd_pcm_close(s->output_unit);
     s->output_unit = NULL;
     return SA_ERROR_NOT_SUPPORTED;
   }
   
   return SA_SUCCESS;
 }
 
