diff --git a/media/libsydneyaudio/src/sydney_audio_mac.c b/media/libsydneyaudio/src/sydney_audio_mac.c
--- a/media/libsydneyaudio/src/sydney_audio_mac.c
+++ b/media/libsydneyaudio/src/sydney_audio_mac.c
@@ -29,16 +29,17 @@
  * and other provisions required by the GPL or the LGPL. If you do not delete
  * the provisions above, a recipient may use your version of this file under
  * the terms of any one of the MPL, the GPL or the LGPL.
  *
  * ***** END LICENSE BLOCK ***** *
  */
 
 #include <pthread.h>
+#include <CoreServices/CoreServices.h>
 #include <AudioUnit/AudioUnit.h>
 #include "sydney_audio.h"
 
 /*
  * The Mac's audio interface is based on a "pull" I/O model, which means you
  * can't just provide a data buffer and tell the audio device to play; you must
  * register a callback and provide data as the device asks for it. To support
  * sydney audio's "write-to-play" style interface, we have to buffer up the
