diff -Nru gdb-7.12/debian/changelog gdb-7.12/debian/changelog
--- gdb-7.12/debian/changelog	2017-01-19 19:28:25.000000000 +0900
+++ gdb-7.12/debian/changelog	2018-02-07 15:36:15.000000000 +0900
@@ -1,3 +1,16 @@
+gdb (7.12-6.deb7moz1) wheezy; urgency=medium
+
+  * Mozilla backport for wheezy.
+  * debian/rules, debian/control: Don't use libbabeltrace, which is not
+    available on wheezy.
+  * debian/rules:
+    - Use DEB_HOST_{ARCH,GNU_TYPE} instead of DEB_TARGET_{ARCH,GNU_TYPE}
+      because the latter is not available on wheezy.
+    - Don't pass --dbgsym-migration to dh_strip, it's not supported on
+      wheezy's debhelper.
+
+ -- Mike Hommey <glandium@mozilla.com>  Wed, 07 Feb 2018 15:36:15 +0900
+
 gdb (7.12-6) unstable; urgency=medium
 
   * debian/patches: import 7.12 branch fixes
diff -Nru gdb-7.12/debian/control gdb-7.12/debian/control
--- gdb-7.12/debian/control	2017-01-17 20:01:41.000000000 +0900
+++ gdb-7.12/debian/control	2018-02-07 15:36:06.000000000 +0900
@@ -33,8 +33,6 @@
                libreadline-dev,
                zlib1g-dev,
                liblzma-dev,
-               libbabeltrace-dev [amd64 armel armhf i386 kfreebsd-amd64 kfreebsd-i386 mips mipsel mips64el powerpc s390x],
-               libbabeltrace-ctf-dev [amd64 armel armhf i386 kfreebsd-amd64 kfreebsd-i386 mips mipsel mips64el powerpc s390x],
                python-dev,
 # Python3 build
                python3-dev,
diff -Nru gdb-7.12/debian/rules gdb-7.12/debian/rules
--- gdb-7.12/debian/rules	2016-12-15 09:31:54.000000000 +0900
+++ gdb-7.12/debian/rules	2018-02-07 15:36:15.000000000 +0900
@@ -4,7 +4,6 @@
 
 DEB_BUILDDIR := $(ALL_BUILDDIR)/objdir
 DEB_DH_INSTALL_SOURCEDIR := $(shell pwd)/debian/tmp
-DEB_DH_STRIP_ARGS_gdb = --dbgsym-migration='gdb-dbg (<< 7.12-1~)'
 
 # Override CDBS's default CFLAGS, which also includes -Wall; gdb
 # does not handle -Wunused well with -Werror, but defaults to
@@ -35,8 +34,8 @@
 DEB_HOST_GNU_SYSTEM := $(shell dpkg-architecture -qDEB_HOST_GNU_SYSTEM)
 DEB_HOST_GNU_CPU := $(shell dpkg-architecture -qDEB_HOST_GNU_CPU)
 DEB_HOST_GNU_TYPE := $(shell dpkg-architecture -qDEB_HOST_GNU_TYPE)
-DEB_TARGET_ARCH := $(shell dpkg-architecture -qDEB_TARGET_ARCH)
-DEB_TARGET_GNU_TYPE := $(shell dpkg-architecture -qDEB_TARGET_GNU_TYPE)
+DEB_TARGET_ARCH := $(shell dpkg-architecture -qDEB_HOST_ARCH)
+DEB_TARGET_GNU_TYPE := $(shell dpkg-architecture -qDEB_HOST_GNU_TYPE)
 
 # Cross configuration support.  Check for an environment variable
 # $GDB_TARGET, or a file debian/target.
@@ -89,10 +88,6 @@
   arch_config_args := --with-libunwind-ia64
 endif
 
-ifneq (,$(filter $(DEB_HOST_ARCH),amd64 armel armhf i386 kfreebsd-amd64 kfreebsd-i386 mips mipsel powerpc s390x))
-  arch_config_args += --with-babeltrace
-endif
-
 ifdef GDB_TARGET
   run_tests := no
 
