diff -Nru glib2.0-2.38.2/debian/changelog glib2.0-2.38.2/debian/changelog
--- glib2.0-2.38.2/debian/changelog	2014-02-15 21:47:00.000000000 +0900
+++ glib2.0-2.38.2/debian/changelog	2018-11-02 09:47:49.000000000 +0900
@@ -1,3 +1,10 @@
+glib2.0 (2.38.2-5.deb7moz1) wheezy; urgency=medium
+
+  * Mozilla backport for Wheezy.
+  * debian/control: Remove :any from dependencies because it's not supported.
+
+ -- Mike Hommey <glandium@mozilla.com>  Fri, 02 Nov 2018 09:47:49 +0900
+
 glib2.0 (2.38.2-5) unstable; urgency=medium
 
   * Upload to unstable.
diff -Nru glib2.0-2.38.2/debian/control glib2.0-2.38.2/debian/control
--- glib2.0-2.38.2/debian/control	2014-02-15 21:50:26.000000000 +0900
+++ glib2.0-2.38.2/debian/control	2018-11-02 09:47:24.000000000 +0900
@@ -24,7 +24,7 @@
                dbus-x11,
                shared-mime-info,
                xterm,
-               python:any (>= 2.6.6-3~),
+               python (>= 2.6.6-3~),
                python-dbus,
                python-gi,
                libxml2-utils,
