diff --git a/debian/changelog b/debian/changelog
index a2948ae..338ed66 100644
--- a/debian/changelog
+++ b/debian/changelog
@@ -1,3 +1,11 @@
+python3-defaults (3.5.3-1.deb7moz1) wheezy; urgency=medium
+
+  * Mozilla backport for wheezy.
+  * debian/control.in
+    * Remove :any from dependencies because it seems to confuse wheezy
+
+ -- Gregory Szorc <gps@mozilla.com>  Tue, 9 May 2018 15:00:00 -0800
+
 python3-defaults (3.5.3-1) unstable; urgency=medium

   * Bump version to 3.5.3.
diff --git a/debian/control.in b/debian/control.in
index 89e1aa4..e58be76 100644
--- a/debian/control.in
+++ b/debian/control.in
@@ -5,8 +5,8 @@ Maintainer: Matthias Klose <doko@debian.org>
 Uploaders: Piotr O<C5><BC>arowski <piotr@debian.org>, Scott Kitterman <scott@kitterman.com>
 Build-Depends: debhelper (>= 9), @bd_i586@
   lsb-release,
-  python3-minimal:any,
-  python3.5-minimal:any,
+  python3-minimal,
+  python3.5-minimal,
   python3-docutils,
   debiandoc-sgml
 Standards-Version: 3.9.8
