diff --git a/debian/changelog b/debian/changelog
index ba3d92c..8a122c3 100644
--- a/debian/changelog
+++ b/debian/changelog
@@ -1,3 +1,12 @@
+sqlite3 (3.16.2-5.deb7moz1) wheezy; urgency=medium
+
+  * Mozilla backport for wheezy.
+  * debian/control, debian/rules:
+    - Downgrade tcl version from 8.6 to 8.5.
+    - Update tcl library path to reflect lack of multiarch.
+
+ -- Gregory Szorc <gps@mozilla.com>  Tue, 11 Dec 2018 12:13:00 -0700
+
 sqlite3 (3.16.2-5) unstable; urgency=medium
 
   * Backport fix for corruption due to REPLACE in an auto-vacuumed database.
diff --git a/debian/control b/debian/control
index af12a8d..e150d02 100644
--- a/debian/control
+++ b/debian/control
@@ -2,8 +2,8 @@ Source: sqlite3
 Section: devel
 Priority: optional
 Maintainer: Laszlo Boszormenyi (GCS) <gcs@debian.org>
-Build-Depends: dpkg-dev (>= 1.16.1~), debhelper (>= 9), autoconf (>= 2.59), libtool (>= 1.5.2), automake, autotools-dev, chrpath, libreadline-dev, tcl8.6-dev, dh-autoreconf
-Build-Conflicts: tcl8.4, tcl8.4-dev, tcl8.5, tcl8.5-dev
+Build-Depends: dpkg-dev (>= 1.16.1~), debhelper (>= 9), autoconf (>= 2.59), libtool (>= 1.5.2), automake, autotools-dev, chrpath, libreadline-dev, tcl8.5-dev, dh-autoreconf
+Build-Conflicts: tcl8.4, tcl8.4-dev
 Homepage: http://www.sqlite.org/
 Standards-Version: 3.9.8
 
diff --git a/debian/rules b/debian/rules
index 72333f7..198f716 100755
--- a/debian/rules
+++ b/debian/rules
@@ -22,10 +22,10 @@ export DEB_BUILD_GNU_TYPE ?= $(shell dpkg-architecture -qDEB_BUILD_GNU_TYPE)
 
 export LDFLAGS += -Wl,--as-needed
 ifeq ($(DEB_BUILD_GNU_TYPE), $(DEB_HOST_GNU_TYPE))
-  confflags += --build $(DEB_HOST_GNU_TYPE) --with-tcl=/usr/lib/$(DEB_HOST_MULTIARCH)/tcl8.6
+  confflags += --build $(DEB_HOST_GNU_TYPE) --with-tcl=/usr/lib/tcl8.5
   export CROSS_BUILDING=no
 else
-  confflags += --build $(DEB_BUILD_GNU_TYPE) --host $(DEB_HOST_GNU_TYPE) --with-tcl=/usr/lib/$(DEB_HOST_MULTIARCH)/tcl8.6
+  confflags += --build $(DEB_BUILD_GNU_TYPE) --host $(DEB_HOST_GNU_TYPE) --with-tcl=/usr/lib/tcl8.5
   LDFLAGS += -L/usr/lib/$(DEB_HOST_MULTIARCH)
   export CROSS_BUILDING=yes
 endif
@@ -115,7 +115,7 @@ install: build
 ifeq ($(DEB_BUILD_GNU_TYPE), $(DEB_HOST_GNU_TYPE))
 	echo "pkg_mkIndex -verbose $(DESTDIR)/usr/lib/tcltk/sqlite3" | \
 		LD_LIBRARY_PATH=$(DESTDIR)/usr/lib/$(DEB_HOST_MULTIARCH) \
-		tclsh8.6
+		tclsh8.5
 	grep -q 'package ifneeded sqlite3' \
 		$(DESTDIR)/usr/lib/tcltk/sqlite3/pkgIndex.tcl || \
 		(echo "pkgIndex.tcl seems to be wrong" && exit 1)
