diff -u xkeyboard-config-2.5.1/debian/control xkeyboard-config-2.5.1/debian/control
--- xkeyboard-config-2.5.1/debian/control
+++ xkeyboard-config-2.5.1/debian/control
@@ -23,6 +23,7 @@
 Depends: ${misc:Depends}
 Breaks: libx11-6 (<< 2:1.4.3)
 Architecture: all
+Multi-Arch: foreign
 Description: X Keyboard Extension (XKB) configuration data
  This package contains configuration data used by the X Keyboard
  Extension (XKB), which allows selection of keyboard layouts when
diff -u xkeyboard-config-2.5.1/debian/changelog xkeyboard-config-2.5.1/debian/changelog
--- xkeyboard-config-2.5.1/debian/changelog
+++ xkeyboard-config-2.5.1/debian/changelog
@@ -1,3 +1,10 @@
+xkeyboard-config (2.5.1-3.deb7moz1) wheezy; urgency=medium
+
+  * Mozilla backport for Wheezy.
+  * Mark xkb-data as "Multi-Arch: foreign".
+
+ -- Mike Hommey <glandium@mozilla.com>  Tue, 06 Nov 2018 17:13:52 +0900
+
 xkeyboard-config (2.5.1-3) unstable; urgency=low
 
   * Cherry-picks from upstream:
