Index: gfx/skia/include/core/SkPreConfig.h
===================================================================
--- gfx/skia/include/core/SkPreConfig.h	(revision 6724)
+++ gfx/skia/include/core/SkPreConfig.h	(working copy)
@@ -94,7 +94,8 @@
 //////////////////////////////////////////////////////////////////////
 
 #if !defined(SK_CPU_BENDIAN) && !defined(SK_CPU_LENDIAN)
-    #if defined (__ppc__) || defined(__ppc64__)
+    #if defined (__ppc__) || defined(__PPC__) || defined(__ppc64__) \
+        || defined(__PPC64__)
         #define SK_CPU_BENDIAN
     #else
         #define SK_CPU_LENDIAN
