# HG changeset patch
# Parent c8288d0c7a1544a590a0cac9c39397ac10c8a45b
Bug 974900 - Add missing include guards to Skia headers - r=gw280

diff --git a/gfx/skia/trunk/include/images/SkImages.h b/gfx/skia/trunk/include/images/SkImages.h
--- a/gfx/skia/trunk/include/images/SkImages.h
+++ b/gfx/skia/trunk/include/images/SkImages.h
@@ -1,14 +1,19 @@
 /*
  * Copyright 2012 Google Inc.
  *
  * Use of this source code is governed by a BSD-style license that can be
  * found in the LICENSE file.
  */
 
+#ifndef SkImages_DEFINED
+#define SkImages_DEFINED
+
 class SkImages {
 public:
     /**
      * Initializes flattenables in the images project.
      */
     static void InitializeFlattenables();
 };
+
+#endif
diff --git a/gfx/skia/trunk/src/core/SkConvolver.h b/gfx/skia/trunk/src/core/SkConvolver.h
--- a/gfx/skia/trunk/src/core/SkConvolver.h
+++ b/gfx/skia/trunk/src/core/SkConvolver.h
@@ -8,16 +8,18 @@
 #include "SkSize.h"
 #include "SkTypes.h"
 #include "SkTArray.h"
 
 // avoid confusion with Mac OS X's math library (Carbon)
 #if defined(__APPLE__)
 #undef FloatToConvolutionFixed
 #undef ConvolutionFixedToFloat
+#undef FloatToFixed
+#undef FixedToFloat
 #endif
 
 // Represents a filter in one dimension. Each output pixel has one entry in this
 // object for the filter values contributing to it. You build up the filter
 // list by calling AddFilter for each output pixel (in order).
 //
 // We do 2-dimensional convolution by first convolving each row by one
 // SkConvolutionFilter1D, then convolving each column by another one.
diff --git a/gfx/skia/trunk/src/gpu/GrAAConvexPathRenderer.h b/gfx/skia/trunk/src/gpu/GrAAConvexPathRenderer.h
--- a/gfx/skia/trunk/src/gpu/GrAAConvexPathRenderer.h
+++ b/gfx/skia/trunk/src/gpu/GrAAConvexPathRenderer.h
@@ -3,24 +3,28 @@
  * Copyright 2012 Google Inc.
  *
  * Use of this source code is governed by a BSD-style license that can be
  * found in the LICENSE file.
  */
 
 #include "GrPathRenderer.h"
 
+#ifndef GrAAConvexPathRenderer_DEFINED
+#define GrAAConvexPathRenderer_DEFINED
 
 class GrAAConvexPathRenderer : public GrPathRenderer {
 public:
     GrAAConvexPathRenderer();
 
     virtual bool canDrawPath(const SkPath& path,
                              const SkStrokeRec& stroke,
                              const GrDrawTarget* target,
                              bool antiAlias) const SK_OVERRIDE;
 
 protected:
     virtual bool onDrawPath(const SkPath& path,
                             const SkStrokeRec& stroke,
                             GrDrawTarget* target,
                             bool antiAlias) SK_OVERRIDE;
 };
+
+#endif
diff --git a/gfx/skia/trunk/src/gpu/GrReducedClip.h b/gfx/skia/trunk/src/gpu/GrReducedClip.h
--- a/gfx/skia/trunk/src/gpu/GrReducedClip.h
+++ b/gfx/skia/trunk/src/gpu/GrReducedClip.h
@@ -1,16 +1,19 @@
 
 /*
  * Copyright 2012 Google Inc.
  *
  * Use of this source code is governed by a BSD-style license that can be
  * found in the LICENSE file.
  */
 
+#ifndef GrReducedClip_DEFINED
+#define GrReducedClip_DEFINED
+
 #include "SkClipStack.h"
 #include "SkTLList.h"
 
 namespace GrReducedClip {
 
 typedef SkTLList<SkClipStack::Element> ElementList;
 
 enum InitialState {
@@ -36,8 +39,10 @@ SK_API void ReduceClipStack(const SkClip
                             const SkIRect& queryBounds,
                             ElementList* result,
                             int32_t* resultGenID,
                             InitialState* initialState,
                             SkIRect* tighterBounds = NULL,
                             bool* requiresAA = NULL);
 
 } // namespace GrReducedClip
+
+#endif
diff --git a/gfx/skia/trunk/src/pathops/SkLineParameters.h b/gfx/skia/trunk/src/pathops/SkLineParameters.h
--- a/gfx/skia/trunk/src/pathops/SkLineParameters.h
+++ b/gfx/skia/trunk/src/pathops/SkLineParameters.h
@@ -1,14 +1,18 @@
 /*
  * Copyright 2012 Google Inc.
  *
  * Use of this source code is governed by a BSD-style license that can be
  * found in the LICENSE file.
  */
+
+#ifndef SkLineParameters_DEFINED
+#define SkLineParameters_DEFINED
+
 #include "SkPathOpsCubic.h"
 #include "SkPathOpsLine.h"
 #include "SkPathOpsQuad.h"
 
 // Sources
 // computer-aided design - volume 22 number 9 november 1990 pp 538 - 549
 // online at http://cagd.cs.byu.edu/~tom/papers/bezclip.pdf
 
@@ -164,8 +168,10 @@ public:
         return -a;
     }
 
 private:
     double a;
     double b;
     double c;
 };
+
+#endif
