error[E0367]: `Drop` impl requires `T: Unpin` but the struct it is implemented for does not
 --> $DIR/conditional-drop-impl.rs:9:9
  |
9 | impl<T: Unpin> Drop for DropImpl<T> {
  |         ^^^^^
  |
note: the implementor must specify the same requirement
 --> $DIR/conditional-drop-impl.rs:5:1
  |
5 | / struct DropImpl<T> {
6 | |     f: T,
7 | | }
  | |_^

error[E0367]: `Drop` impl requires `T: Unpin` but the struct it is implemented for does not
  --> $DIR/conditional-drop-impl.rs:14:1
   |
14 | / pin_project! {
15 | |     //~^ ERROR E0367
16 | |     struct PinnedDropImpl<T> {
17 | |         #[pin]
...  |
23 | |     }
24 | | }
   | |_^
   |
note: the implementor must specify the same requirement
  --> $DIR/conditional-drop-impl.rs:14:1
   |
14 | / pin_project! {
15 | |     //~^ ERROR E0367
16 | |     struct PinnedDropImpl<T> {
17 | |         #[pin]
...  |
23 | |     }
24 | | }
   | |_^
   = note: this error originates in the macro `$crate::__pin_project_internal` (in Nightly builds, run with -Z macro-backtrace for more info)
