diff --git a/media/libstagefright/binding/mp4parse/lib.rs b/media/libstagefright/binding/mp4parse/lib.rs
index af62402..7c8e382 100644
--- a/media/libstagefright/binding/mp4parse/lib.rs
+++ b/media/libstagefright/binding/mp4parse/lib.rs
@@ -4,7 +4,7 @@
 // License, v. 2.0. If a copy of the MPL was not distributed with this
 // file, You can obtain one at https://mozilla.org/MPL/2.0/.
 
-extern crate byteorder;
+mod byteorder; // 'extern crate' upstream.
 use byteorder::ReadBytesExt;
 use std::error::Error as ErrorTrait; // For Err(e) => e.description().
 use std::io::{Read, BufRead, Take};
