diff -ru a/configure.in b/configure.in
--- a/configure.in	2011-01-17 23:34:07.000000000 -0500
+++ b/configure.in	2012-01-25 20:40:27.919485606 -0500
@@ -841,14 +841,6 @@
 LIBC_PROG_BINUTILS
 AC_SUBST(MIG)dnl Needed by sysdeps/mach/configure.in
 
-# Accept binutils 2.13 or newer.
-AC_CHECK_PROG_VER(AS, $AS, --version,
-		  [GNU assembler.* \([0-9]*\.[0-9.]*\)],
-		  [2.1[3-9]*], AS=: critic_missing="$critic_missing as")
-AC_CHECK_PROG_VER(LD, $LD, --version,
-		  [GNU ld.* \([0-9][0-9]*\.[0-9.]*\)],
-		  [2.1[3-9]*], LD=: critic_missing="$critic_missing ld")
-
 # We need the physical current working directory.  We cannot use the
 # "pwd -P" shell builtin since that's not portable.  Instead we try to
 # find a pwd binary.  Note that assigning to the PWD environment
@@ -2175,6 +2167,7 @@
 fi
 AC_SUBST(old_glibc_headers)
 
+libc_cv_slibdir=${prefix}/lib64
 AC_SUBST(libc_cv_slibdir)
 AC_SUBST(libc_cv_localedir)
 AC_SUBST(libc_cv_sysconfdir)
diff -ru a/csu/Makefile b/csu/Makefile
--- a/csu/Makefile	2011-01-17 23:34:07.000000000 -0500
+++ b/csu/Makefile	2012-01-23 13:58:28.957792633 -0500
@@ -223,8 +223,7 @@
 		   if [ -z "$$os" ]; then \
 		     os=Linux; \
 		   fi; \
-		   printf '"Compiled on a %s %s system on %s.\\n"\n' \
-			  "$$os" "$$version" "`date +%Y-%m-%d`";; \
+                   ;; \
 	   *) ;; \
 	 esac; \
 	 files="$(all-Banner-files)";				\
diff -ru a/elf/Makefile b/elf/Makefile
--- a/elf/Makefile	2008-10-31 16:35:11.000000000 -0400
+++ b/elf/Makefile	2012-02-16 12:20:00.038593752 -0500
@@ -295,18 +295,11 @@
 z-now-yes = -Wl,-z,now
 
 $(objpfx)ld.so: $(objpfx)librtld.os $(ld-map)
-	@rm -f $@.lds
-	$(LINK.o) -nostdlib -nostartfiles -shared $(z-now-$(bind-now))	\
-		  $(LDFLAGS-rtld) -Wl,-z,defs -Wl,--verbose 2>&1 |	\
-		  LC_ALL=C \
-		  sed -e '/^=========/,/^=========/!d;/^=========/d'	\
-		      -e 's/\. = 0 + SIZEOF_HEADERS;/& _begin = . - SIZEOF_HEADERS;/' \
-		  > $@.lds
 	$(LINK.o) -nostdlib -nostartfiles -shared -o $@			\
 		  $(LDFLAGS-rtld) -Wl,-z,defs $(z-now-$(bind-now))	\
 		  $(filter-out $(map-file),$^) $(load-map-file)		\
-		  -Wl,-soname=$(rtld-installed-name) -T $@.lds
-	rm -f $@.lds
+		  -Wl,-soname=$(rtld-installed-name)			\
+		  -Wl,-defsym=_begin=0
 
 # interp.c exists just to get this string into the libraries.
 CFLAGS-interp.c = -D'RUNTIME_LINKER="$(slibdir)/$(rtld-installed-name)"' \
diff -ru a/localedata/Makefile b/localedata/Makefile
--- a/localedata/Makefile	2006-04-26 01:14:03.000000000 -0400
+++ b/localedata/Makefile	2012-02-17 10:31:24.592345047 -0500
@@ -113,7 +113,7 @@
 	$(make-target-directory)
 	rm -f $(@:.gz=) $@
 	$(INSTALL_DATA) $< $(@:.gz=)
-	gzip -9 $(@:.gz=)
+	gzip -9n $(@:.gz=)
 
 # Install the locale source files in the appropriate directory.
 $(inst_i18ndir)/locales/%: locales/% $(+force); $(do-install)
diff -ru a/Makeconfig b/Makeconfig
--- a/Makeconfig	2006-07-10 17:42:27.000000000 -0400
+++ b/Makeconfig	2012-02-17 08:28:31.859584817 -0500
@@ -674,7 +674,7 @@
 	   $(foreach lib,$(libof-$(basename $(@F))) \
 		         $(libof-$(<F)) $(libof-$(@F)),$(CPPFLAGS-$(lib))) \
 	   $(CPPFLAGS-$(<F)) $(CPPFLAGS-$(@F)) $(CPPFLAGS-$(basename $(@F)))
-override CFLAGS	= -std=gnu99 \
+override CFLAGS	= -std=gnu99 -fgnu89-inline \
 		  $(filter-out %frame-pointer,$(+cflags)) $(+gccwarn-c) \
 		  $(sysdep-CFLAGS) $(CFLAGS-$(suffix $@)) $(CFLAGS-$(<F)) \
 		  $(CFLAGS-$(@F))
diff -ru a/Makerules b/Makerules
--- a/Makerules	2011-01-17 23:34:07.000000000 -0500
+++ b/Makerules	2012-01-30 08:47:56.565068903 -0500
@@ -977,9 +977,9 @@
 	 echo '   Use the shared library, but some functions are only in';\
 	 echo '   the static library, so try that secondarily.  */';\
 	 cat $<; \
-	 echo 'GROUP ( $(slibdir)/libc.so$(libc.so-version)' \
-	      '$(libdir)/$(patsubst %,$(libtype.oS),$(libprefix)$(libc-name))'\
-	      ' AS_NEEDED (' $(slibdir)/$(rtld-installed-name) ') )' \
+	 echo 'GROUP ( libc.so$(libc.so-version)' \
+	      '$(patsubst %,$(libtype.oS),$(libprefix)$(libc-name))'\
+	      ' AS_NEEDED (' $(rtld-installed-name) ') )' \
 	) > $@.new
 	mv -f $@.new $@
 
diff -ru a/nscd/nscd_stat.c b/nscd/nscd_stat.c
--- a/nscd/nscd_stat.c	2011-01-17 23:34:07.000000000 -0500
+++ b/nscd/nscd_stat.c	2012-01-23 15:54:45.231607606 -0500
@@ -38,7 +38,7 @@
 
 
 /* We use this to make sure the receiver is the same.  */
-static const char compilation[21] = __DATE__ " " __TIME__;
+static const char compilation[21] = "don't need this";
 
 /* Statistic data for one database.  */
 struct dbstat
