diff -ruN a/ltmain.sh b/ltmain.sh
--- a/ltmain.sh	2009-12-05 12:18:53.000000000 -0500
+++ b/ltmain.sh	2012-05-07 16:19:31.871827967 -0400
@@ -2932,7 +2932,7 @@
         func_extract_an_archive "$my_xdir" "$my_xabs"
 	;;
       esac
-      my_oldobjs="$my_oldobjs "`find $my_xdir -name \*.$objext -print -o -name \*.lo -print | $NL2SP`
+      my_oldobjs="$my_oldobjs "`find $my_xdir -name \*.$objext -print -o -name \*.lo -print | sort | $NL2SP`
     done
 
     func_extract_archives_result="$my_oldobjs"
