error: `self` parameter is only allowed in associated functions
  --> tests/ui/pinned_drop/self.rs:17:26
   |
17 |                     fn f(self: ()) {} //~ ERROR `self` parameter is only allowed in associated functions
   |                          ^^^^ not semantically valid as function parameter
...
20 |             t!();
   |             ---- in this macro invocation
   |
   = note: associated functions are those in `impl` or `trait` definitions
   = note: this error originates in the macro `t` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0434]: can't capture dynamic environment in a fn item
  --> tests/ui/pinned_drop/self.rs:15:29
   |
15 |                     let _ = self; //~ ERROR E0434
   |                             ^^^^
...
20 |             t!();
   |             ---- in this macro invocation
   |
   = help: use the `|| { ... }` closure form instead
   = note: this error originates in the macro `t` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0423]: expected value, found struct `S`
  --> tests/ui/pinned_drop/self.rs:38:27
   |
30 | /     pub struct S {
31 | |         x: (),
32 | |     }
   | |_____- `S` defined here
...
38 |               let _: Self = Self; //~ ERROR E0423
   |                             ^^^^ help: use struct literal syntax instead: `S { x: val }`

error[E0308]: mismatched types
  --> tests/ui/pinned_drop/self.rs:37:25
   |
37 |             let _: () = self; //~ ERROR E0308
   |                    --   ^^^^ expected `()`, found struct `Pin`
   |                    |
   |                    expected due to this
   |
   = note: expected unit type `()`
                 found struct `Pin<&mut S>`

error[E0308]: mismatched types
  --> tests/ui/pinned_drop/self.rs:50:25
   |
50 |             let _: () = self; //~ ERROR E0308
   |                    --   ^^^^ expected `()`, found struct `Pin`
   |                    |
   |                    expected due to this
   |
   = note: expected unit type `()`
                 found struct `Pin<&mut E>`

error[E0533]: expected unit struct, unit variant or constant, found struct variant `Self::V`
  --> tests/ui/pinned_drop/self.rs:51:27
   |
51 |             let _: Self = Self::V; //~ ERROR E0533
   |                           ^^^^^^^
