diff --git a/src/channel/PhutilSocketChannel.php b/src/channel/PhutilSocketChannel.php
index 4bd2a47a..2ad3fd85 100644
--- a/src/channel/PhutilSocketChannel.php
+++ b/src/channel/PhutilSocketChannel.php
@@ -46,9 +46,11 @@ final class PhutilSocketChannel extends PhutilChannel {
       if (!$socket) {
         continue;
       }
-      $ok = stream_set_blocking($socket, false);
-      if (!$ok) {
-        throw new Exception(pht('Failed to set socket nonblocking!'));
+      if (strtoupper(substr(PHP_OS, 0, 3)) !== 'WIN') {
+        $ok = stream_set_blocking($socket, false);
+        if (!$ok) {
+          throw new Exception(pht('Failed to set socket nonblocking!'));
+        }
       }
     }
 
