diff --git a/src/workflow/ArcanistDiffWorkflow.php b/src/workflow/ArcanistDiffWorkflow.php
index 38aa4b62..c9b7e215 100644
--- a/src/workflow/ArcanistDiffWorkflow.php
+++ b/src/workflow/ArcanistDiffWorkflow.php
@@ -899,7 +899,7 @@ EOTEXT
       throw new Exception(pht('Repository API is not supported.'));
     }
 
-    $limit = 1024 * 1024 * 4;
+    $limit = 1024 * 1024 * 12;
     foreach ($changes as $change) {
       $size = 0;
       foreach ($change->getHunks() as $hunk) {
