diff --git a/mozmill/mozmill/extension/resource/stdlib/httpd.js b/mozmill/mozmill/extension/resource/stdlib/httpd.js
index cf974b3..35a8d36 100644
--- a/mozmill/mozmill/extension/resource/stdlib/httpd.js
+++ b/mozmill/mozmill/extension/resource/stdlib/httpd.js
@@ -48,6 +48,14 @@
 
 Components.utils.import("resource://gre/modules/XPCOMUtils.jsm");
 
+var EXPORTED_SYMBOLS = ['getServer'];
+
+/**
+ * Overwrite both dump functions because we do not wanna have this output for Mozmill
+ */
+function dump() {}
+function dumpn() {}
+
 const Cc = Components.classes;
 const Ci = Components.interfaces;
 const Cr = Components.results;
@@ -5173,3 +5181,20 @@ function server(port, basePath)
 
   DEBUG = false;
 }
+
+function getServer (port, basePath) {
+  if (basePath) {
+    var lp = Cc["@mozilla.org/file/local;1"]
+               .createInstance(Ci.nsILocalFile);
+    lp.initWithPath(basePath);
+   }
+
+   var srv = new nsHttpServer();
+   if (lp)
+     srv.registerDirectory("/", lp);
+   srv.registerContentType("sjs", SJS_TYPE);
+   srv.identity.setPrimary("http", "localhost", port);
+   srv._port = port;
+
+   return srv;
+}
