*** Source:appleevents.r.2	Tue Aug 12 12:00:31 1997
--- Source:appleevents.r.3	Fri Mar 13 03:29:55 1998
***************
*** 80,84 ****
--- 80,96 ----
  #define kAEHighPriority 				0x00000001			/*  post message at the front of the event queue (same as nAttnMsg)  */
  
  
+ /*--------------------------aedt ¥ Apple Events Template---------------------------------*/
+ /* Resource definition used for associating a value with an apple event */
+ /* This really only useful for general dispatching */
+ 
+ type 'aedt' {
+ 	wide array {
+ 	unsigned longint;	/* Event Class 	*/
+ 	unsigned longint;	/* Event ID 	*/
+ 	unsigned longint;	/* Value 	*/
+ 	};
+ };
+ 
  #endif /* __APPLEEVENTS_R__ */
  
