*** menusharing.c	Mon Mar 16 16:42:29 1998
--- menusharing.c.new	Mon Mar 16 16:43:43 1998
***************
*** 110,116 ****
  
  #include <processes.h>
  #include <Menus.h>
! #include <GestaltEqu.h>
  #include "menusharing.h"
  
  #define kAEDontRecord 0x00001000 /*this isn't defined in all versions of <AppleEvents.h>*/
--- 110,116 ----
  
  #include <processes.h>
  #include <Menus.h>
! #include <Gestalt.h>
  #include "menusharing.h"
  
  #define kAEDontRecord 0x00001000 /*this isn't defined in all versions of <AppleEvents.h>*/
