*** Source:stat.mac.h.2	Fri Oct 17 19:21:19 1997
--- Source:stat.mac.h.3	Tue Mar  3 17:55:14 1998
***************
*** 56,61 ****
--- 56,63 ----
  #define S_ISCHR(m)	(((m)&(S_IFMT)) == (S_IFCHR))
  #define S_ISBLK(m)	(((m)&(S_IFMT)) == (S_IFBLK))
  #define S_ISREG(m)	(((m)&(S_IFMT)) == (S_IFREG))
+ #define S_ISLNK(m)	(((m)&(S_IFMT)) == (S_IFLNK))
+ #define S_ISSOCK(m)	(((m)&(S_IFMT)) == (S_IFSOCK))
  
  struct stat
  {
