error: duplicate #[pin] attribute
 --> $DIR/add-attr-to-struct.rs:6:1
  |
6 | #[add_attr(struct)] //~ ERROR duplicate #[pin] attribute
  | ^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: #[pin] attribute may only be used on fields of structs or variants
  --> $DIR/add-attr-to-struct.rs:12:1
   |
12 | #[add_attr(struct)] //~ ERROR #[pin] attribute may only be used on fields of structs or variants
   | ^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)
