error[E0119]: conflicting implementations of trait `std::marker::Unpin` for type `Foo<_, _>`:
  --> $DIR/conflict-unpin.rs:5:1
   |
5  | #[pin_project] //~ ERROR E0119
   | ^^^^^^^^^^^^^^ conflicting implementation for `Foo<_, _>`
...
13 | impl<T, U> Unpin for Foo<T, U> where T: Unpin {} // Conditional Unpin impl
   | --------------------------------------------- first implementation here
   |
   = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0119]: conflicting implementations of trait `std::marker::Unpin` for type `Bar<_, _>`:
  --> $DIR/conflict-unpin.rs:17:1
   |
17 | #[pin_project] //~ ERROR E0119
   | ^^^^^^^^^^^^^^ conflicting implementation for `Bar<_, _>`
...
25 | impl<T, U> Unpin for Bar<T, U> {} // Non-conditional Unpin impl
   | ------------------------------ first implementation here
   |
   = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0119]: conflicting implementations of trait `std::marker::Unpin` for type `Baz<_, _>`:
  --> $DIR/conflict-unpin.rs:27:1
   |
27 | #[pin_project] //~ ERROR E0119
   | ^^^^^^^^^^^^^^ conflicting implementation for `Baz<_, _>`
...
35 | impl<T: Unpin, U: Unpin> Unpin for Baz<T, U> {} // Conditional Unpin impl
   | -------------------------------------------- first implementation here
   |
   = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)
