error[E0119]: conflicting implementations of trait `_::_pin_project::UnsafeUnpin` for type `Foo<_, _>`
  --> tests/ui/pin_project/impl-unsafe-unpin.rs:3:1
   |
3  | #[pin_project] //~ ERROR E0119
   | ^^^^^^^^^^^^^^ conflicting implementation for `Foo<_, _>`
...
10 | unsafe impl<T, U> UnsafeUnpin for Foo<T, U> where T: Unpin {}
   | ------------------------------------------- first implementation here
   |
   = note: this error originates in the derive macro `::pin_project::__private::__PinProjectInternalDerive` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0119]: conflicting implementations of trait `_::_pin_project::UnsafeUnpin` for type `Bar<_, _>`
  --> tests/ui/pin_project/impl-unsafe-unpin.rs:12:1
   |
12 | #[pin_project] //~ ERROR E0119
   | ^^^^^^^^^^^^^^ conflicting implementation for `Bar<_, _>`
...
19 | unsafe impl<T, U> UnsafeUnpin for Bar<T, U> {}
   | ------------------------------------------- first implementation here
   |
   = note: this error originates in the derive macro `::pin_project::__private::__PinProjectInternalDerive` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0119]: conflicting implementations of trait `_::_pin_project::UnsafeUnpin` for type `Baz<_, _>`
  --> tests/ui/pin_project/impl-unsafe-unpin.rs:21:1
   |
21 | #[pin_project] //~ ERROR E0119
   | ^^^^^^^^^^^^^^ conflicting implementation for `Baz<_, _>`
...
28 | unsafe impl<T: Unpin, U: Unpin> UnsafeUnpin for Baz<T, U> {}
   | --------------------------------------------------------- first implementation here
   |
   = note: this error originates in the derive macro `::pin_project::__private::__PinProjectInternalDerive` (in Nightly builds, run with -Z macro-backtrace for more info)
