error: #[pin_project] attribute may not be used on #[repr(packed)] types
 --> tests/ui/pin_project/packed.rs:4:8
  |
4 | #[repr(packed, C)] //~ ERROR may not be used on #[repr(packed)] types
  |        ^^^^^^

error: #[pin_project] attribute may not be used on #[repr(packed)] types
  --> tests/ui/pin_project/packed.rs:11:8
   |
11 | #[repr(packed, C)] //~ ERROR may not be used on #[repr(packed)] types
   |        ^^^^^^

error: #[pin_project] attribute may not be used on #[repr(packed)] types
  --> tests/ui/pin_project/packed.rs:19:8
   |
19 | #[repr(packed(2))] //~ ERROR may not be used on #[repr(packed)] types
   |        ^^^^^^^^^

error: #[pin_project] attribute may not be used on #[repr(packed)] types
  --> tests/ui/pin_project/packed.rs:26:8
   |
26 | #[repr(packed(2))] //~ ERROR may not be used on #[repr(packed)] types
   |        ^^^^^^^^^
