error[E0446]: private type `PrivateEnum` in public interface
 --> $DIR/private_in_public-enum.rs:6:13
  |
6 |     Variant(PrivateEnum), //~ ERROR E0446
  |             ^^^^^^^^^^^ can't leak private type
...
9 | enum PrivateEnum {
  | ---------------- `PrivateEnum` declared as private

error[E0446]: private type `foo::PrivateEnum` in public interface
  --> $DIR/private_in_public-enum.rs:15:17
   |
15 |         Variant(PrivateEnum), //~ ERROR E0446
   |                 ^^^^^^^^^^^ can't leak private type
...
18 |     enum PrivateEnum {
   |     ---------------- `foo::PrivateEnum` declared as private
