error[E0308]: mismatched types
  --> tests/ui/pin_project/remove-attr-from-field.rs:27:38
   |
27 |     let _: Pin<&mut PhantomPinned> = x.field; //~ ERROR E0308
   |            -----------------------   ^^^^^^^ expected struct `Pin`, found `&mut PhantomPinned`
   |            |
   |            expected due to this
   |
   = note:         expected struct `Pin<&mut PhantomPinned>`
           found mutable reference `&mut PhantomPinned`

error[E0308]: mismatched types
  --> tests/ui/pin_project/remove-attr-from-field.rs:31:38
   |
31 |     let _: Pin<&mut PhantomPinned> = x.field; //~ ERROR E0308
   |            -----------------------   ^^^^^^^ expected struct `Pin`, found `&mut PhantomPinned`
   |            |
   |            expected due to this
   |
   = note:         expected struct `Pin<&mut PhantomPinned>`
           found mutable reference `&mut PhantomPinned`
