error: borrow of packed field is unsafe and requires unsafe function or block (error E0133)
  --> $DIR/safe_packed_borrows.rs:21:5
   |
21 |     &a.f; //~ ERROR borrow of packed field is unsafe and requires unsafe function or block
   |     ^^^^
   |
note: the lint level is defined here
  --> $DIR/safe_packed_borrows.rs:1:11
   |
1  | #![forbid(safe_packed_borrows)]
   |           ^^^^^^^^^^^^^^^^^^^
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #46043 <https://github.com/rust-lang/rust/issues/46043>
   = note: fields of packed structs might be misaligned: dereferencing a misaligned pointer or even just creating a misaligned reference is undefined behavior

error: borrow of packed field is unsafe and requires unsafe function or block (error E0133)
  --> $DIR/safe_packed_borrows.rs:22:13
   |
22 |     let _ = &a.f; //~ ERROR borrow of packed field is unsafe and requires unsafe function or block
   |             ^^^^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #46043 <https://github.com/rust-lang/rust/issues/46043>
   = note: fields of packed structs might be misaligned: dereferencing a misaligned pointer or even just creating a misaligned reference is undefined behavior

error: borrow of packed field is unsafe and requires unsafe function or block (error E0133)
  --> $DIR/safe_packed_borrows.rs:25:5
   |
25 |     &b.f; //~ ERROR borrow of packed field is unsafe and requires unsafe function or block
   |     ^^^^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #46043 <https://github.com/rust-lang/rust/issues/46043>
   = note: fields of packed structs might be misaligned: dereferencing a misaligned pointer or even just creating a misaligned reference is undefined behavior

error: borrow of packed field is unsafe and requires unsafe function or block (error E0133)
  --> $DIR/safe_packed_borrows.rs:26:13
   |
26 |     let _ = &b.f; //~ ERROR borrow of packed field is unsafe and requires unsafe function or block
   |             ^^^^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #46043 <https://github.com/rust-lang/rust/issues/46043>
   = note: fields of packed structs might be misaligned: dereferencing a misaligned pointer or even just creating a misaligned reference is undefined behavior
