error: reference to packed field is unaligned
  --> $DIR/unaligned_references.rs:18:5
   |
18 |     &a.f; //~ ERROR reference to packed field is unaligned
   |     ^^^^
   |
note: the lint level is defined here
  --> $DIR/unaligned_references.rs:1:11
   |
1  | #![forbid(unaligned_references)]
   |           ^^^^^^^^^^^^^^^^^^^^
   = note: fields of packed structs are not properly aligned, and creating a misaligned reference is undefined behavior (even if that reference is never dereferenced)

error: reference to packed field is unaligned
  --> $DIR/unaligned_references.rs:19:13
   |
19 |     let _ = &a.f; //~ ERROR reference to packed field is unaligned
   |             ^^^^
   |
   = note: fields of packed structs are not properly aligned, and creating a misaligned reference is undefined behavior (even if that reference is never dereferenced)

error: reference to packed field is unaligned
  --> $DIR/unaligned_references.rs:22:5
   |
22 |     &b.f; //~ ERROR reference to packed field is unaligned
   |     ^^^^
   |
   = note: fields of packed structs are not properly aligned, and creating a misaligned reference is undefined behavior (even if that reference is never dereferenced)

error: reference to packed field is unaligned
  --> $DIR/unaligned_references.rs:23:13
   |
23 |     let _ = &b.f; //~ ERROR reference to packed field is unaligned
   |             ^^^^
   |
   = note: fields of packed structs are not properly aligned, and creating a misaligned reference is undefined behavior (even if that reference is never dereferenced)
