--- sources.mozbuild-	2015-12-31 01:09:35.663118019 -0800
+++ sources.mozbuild	2015-12-31 01:12:04.970093876 -0800
@@ -2,8 +2,6 @@
 celt_sources = [
     'celt/bands.c',
     'celt/celt.c',
-    'celt/celt_decoder.c',
-    'celt/celt_encoder.c',
     'celt/celt_lpc.c',
     'celt/cwrs.c',
     'celt/entcode.c',
@@ -20,6 +18,13 @@
     'celt/vq.c',
 ]
 
+opus_nonunified_sources = [
+    'celt/celt_decoder.c',
+    'celt/celt_encoder.c',
+    'silk/LPC_inv_pred_gain.c',
+    'silk/NLSF2A.c',
+]
+
 celt_sources_sse = [
     'celt/x86/pitch_sse.c',
     'celt/x86/x86_celt_map.c',
@@ -105,8 +110,6 @@
     'silk/log2lin.c',
     'silk/LP_variable_cutoff.c',
     'silk/LPC_analysis_filter.c',
-    'silk/LPC_inv_pred_gain.c',
-    'silk/NLSF2A.c',
     'silk/NLSF_decode.c',
     'silk/NLSF_del_dec_quant.c',
     'silk/NLSF_encode.c',
