error[E0119]: conflicting implementations of trait `std::marker::Unpin` for type `Foo<_, _>`:
  --> $DIR/conflict-unpin.rs:5:1
   |
5  | / pin_project! { //~ ERROR E0119
6  | |     struct Foo<T, U> {
7  | |         #[pin]
8  | |         future: T,
9  | |         field: U,
10 | |     }
11 | | }
   | |_^ conflicting implementation for `Foo<_, _>`
...
14 |   impl<T, U> Unpin for Foo<T, U> where T: Unpin {} // Conditional Unpin impl
   |   --------------------------------------------- first implementation here
   |
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0119]: conflicting implementations of trait `std::marker::Unpin` for type `Bar<_, _>`:
  --> $DIR/conflict-unpin.rs:18:1
   |
18 | / pin_project! { //~ ERROR E0119
19 | |     struct Bar<T, U> {
20 | |         #[pin]
21 | |         future: T,
22 | |         field: U,
23 | |     }
24 | | }
   | |_^ conflicting implementation for `Bar<_, _>`
...
27 |   impl<T, U> Unpin for Bar<T, U> {} // Non-conditional Unpin impl
   |   ------------------------------ first implementation here
   |
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0119]: conflicting implementations of trait `std::marker::Unpin` for type `Baz<_, _>`:
  --> $DIR/conflict-unpin.rs:29:1
   |
29 | / pin_project! { //~ ERROR E0119
30 | |     struct Baz<T, U> {
31 | |         #[pin]
32 | |         future: T,
33 | |         field: U,
34 | |     }
35 | | }
   | |_^ conflicting implementation for `Baz<_, _>`
...
38 |   impl<T: Unpin, U: Unpin> Unpin for Baz<T, U> {} // Conditional Unpin impl
   |   -------------------------------------------- first implementation here
   |
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
