error: no rules expected the token `[`
 --> $DIR/invalid.rs:3:1
  |
3 | / pin_project! {
4 | |     struct A<T> {
5 | |         #[pin()] //~ ERROR no rules expected the token `(`
6 | |         pinned: T,
7 | |     }
8 | | }
  | |_^ no rules expected this token in macro call
  |
  = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: no rules expected the token `[`
  --> $DIR/invalid.rs:17:1
   |
17 | / pin_project! {
18 | |     struct C<T> {
19 | |         #[pin]
20 | |         #[pin] //~ ERROR no rules expected the token `#`
21 | |         pinned: T,
22 | |     }
23 | | }
   | |_^ no rules expected this token in macro call
   |
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find attribute `pin` in this scope
  --> $DIR/invalid.rs:11:7
   |
11 |     #[pin] //~ ERROR cannot find attribute `pin` in this scope
   |       ^^^
