error: no rules expected the token `[`
 --> $DIR/invalid-bounds.rs:3:1
  |
3 | / pin_project! {
4 | |     struct Generics1<T: 'static : Sized> { //~ ERROR no rules expected the token `:`
5 | |         field: T,
6 | |     }
7 | | }
  | |_^ no rules expected this token in macro call
  |
  = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: no rules expected the token `[`
  --> $DIR/invalid-bounds.rs:9:1
   |
9  | / pin_project! {
10 | |     struct Generics2<T: 'static : ?Sized> { //~ ERROR no rules expected the token `:`
11 | |         field: T,
12 | |     }
13 | | }
   | |_^ no rules expected this token in macro call
   |
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected one of `+`, `,`, `=`, or `>`, found `:`
  --> $DIR/invalid-bounds.rs:15:1
   |
15 | / pin_project! {
16 | |     struct Generics3<T: Sized : 'static> { //~ ERROR expected one of `+`, `,`, `=`, or `>`, found `:`
17 | |         field: T,
18 | |     }
19 | | }
   | | ^
   | | |
   | | expected one of `+`, `,`, `=`, or `>`
   | |_unexpected token
   |   in this macro invocation
   |
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected one of `+`, `,`, `=`, or `>`, found `:`
  --> $DIR/invalid-bounds.rs:21:1
   |
21 | / pin_project! {
22 | |     struct Generics4<T: ?Sized : 'static> { //~ ERROR expected one of `+`, `,`, `=`, or `>`, found `:`
23 | |         field: T,
24 | |     }
25 | | }
   | | ^
   | | |
   | | expected one of `+`, `,`, `=`, or `>`
   | |_unexpected token
   |   in this macro invocation
   |
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected one of `+`, `,`, `=`, or `>`, found `:`
  --> $DIR/invalid-bounds.rs:27:1
   |
27 | / pin_project! {
28 | |     struct Generics5<T: Sized : ?Sized> { //~ ERROR expected one of `+`, `,`, `=`, or `>`, found `:`
29 | |         field: T,
30 | |     }
31 | | }
   | | ^
   | | |
   | | expected one of `+`, `,`, `=`, or `>`
   | |_unexpected token
   |   in this macro invocation
   |
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: no rules expected the token `[`
  --> $DIR/invalid-bounds.rs:33:1
   |
33 | / pin_project! {
34 | |     struct Generics6<T: ?Sized : Sized> { //~ ERROR no rules expected the token `Sized`
35 | |         field: T,
36 | |     }
37 | | }
   | |_^ no rules expected this token in macro call
   |
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: no rules expected the token `[`
  --> $DIR/invalid-bounds.rs:39:1
   |
39 | / pin_project! {
40 | |     struct WhereClause1<T>
41 | |     where
42 | |         T: 'static : Sized //~ ERROR no rules expected the token `:`
...  |
45 | |     }
46 | | }
   | |_^ no rules expected this token in macro call
   |
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: no rules expected the token `[`
  --> $DIR/invalid-bounds.rs:48:1
   |
48 | / pin_project! {
49 | |     struct WhereClause2<T>
50 | |     where
51 | |         T: 'static : ?Sized //~ ERROR no rules expected the token `:`
...  |
54 | |     }
55 | | }
   | |_^ no rules expected this token in macro call
   |
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected `where`, or `{` after struct name, found `:`
  --> $DIR/invalid-bounds.rs:57:1
   |
57 | / pin_project! {
58 | |     struct WhereClause3<T>
59 | |     where
60 | |         T: Sized : 'static //~ ERROR expected `where`, or `{` after struct name, found `:`
...  |
63 | |     }
64 | | }
   | | ^
   | | |
   | |_expected `where`, or `{` after struct name
   |   in this macro invocation
   |
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected `where`, or `{` after struct name, found `:`
  --> $DIR/invalid-bounds.rs:66:1
   |
66 | / pin_project! {
67 | |     struct WhereClause4<T>
68 | |     where
69 | |         T: ?Sized : 'static //~ ERROR expected `where`, or `{` after struct name, found `:`
...  |
72 | |     }
73 | | }
   | | ^
   | | |
   | |_expected `where`, or `{` after struct name
   |   in this macro invocation
   |
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected `where`, or `{` after struct name, found `:`
  --> $DIR/invalid-bounds.rs:75:1
   |
75 | / pin_project! {
76 | |     struct WhereClause5<T>
77 | |     where
78 | |         T: Sized : ?Sized //~ ERROR expected `where`, or `{` after struct name, found `:`
...  |
81 | |     }
82 | | }
   | | ^
   | | |
   | |_expected `where`, or `{` after struct name
   |   in this macro invocation
   |
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: no rules expected the token `[`
  --> $DIR/invalid-bounds.rs:84:1
   |
84 | / pin_project! {
85 | |     struct WhereClause6<T>
86 | |     where
87 | |         T: ?Sized : Sized //~ ERROR no rules expected the token `Sized`
...  |
90 | |     }
91 | | }
   | |_^ no rules expected this token in macro call
   |
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
