error[E0496]: lifetime name `'__pin` shadows a lifetime name that is already in scope
 --> $DIR/overlapping_lifetime_names.rs:3:1
  |
3 | / pin_project! { //~ ERROR E0496
4 | |     pub struct Foo<'__pin, T> { //~ ERROR E0263
  | |                    ------ first declared here
5 | |         #[pin]
6 | |         field: &'__pin mut T,
7 | |     }
8 | | }
  | |_^ lifetime `'__pin` already in scope
  |
  = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0496]: lifetime name `'__pin` shadows a lifetime name that is already in scope
 --> $DIR/overlapping_lifetime_names.rs:3:1
  |
3 | / pin_project! { //~ ERROR E0496
4 | |     pub struct Foo<'__pin, T> { //~ ERROR E0263
  | |                    ------ first declared here
5 | |         #[pin]
6 | |         field: &'__pin mut T,
7 | |     }
8 | | }
  | |_^ lifetime `'__pin` already in scope
  |
  = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0263]: lifetime name `'__pin` declared twice in the same scope
 --> $DIR/overlapping_lifetime_names.rs:4:20
  |
3 | / pin_project! { //~ ERROR E0496
4 | |     pub struct Foo<'__pin, T> { //~ ERROR E0263
  | |                    ^^^^^^ declared twice
5 | |         #[pin]
6 | |         field: &'__pin mut T,
7 | |     }
8 | | }
  | |_- previous declaration here

error[E0263]: lifetime name `'__pin` declared twice in the same scope
 --> $DIR/overlapping_lifetime_names.rs:4:20
  |
3 | / pin_project! { //~ ERROR E0496
4 | |     pub struct Foo<'__pin, T> { //~ ERROR E0263
  | |                    ^^^^^^ declared twice
5 | |         #[pin]
6 | |         field: &'__pin mut T,
7 | |     }
8 | | }
  | |_- previous declaration here

error[E0263]: lifetime name `'__pin` declared twice in the same scope
 --> $DIR/overlapping_lifetime_names.rs:4:20
  |
3 | / pin_project! { //~ ERROR E0496
4 | |     pub struct Foo<'__pin, T> { //~ ERROR E0263
  | |                    ^^^^^^ declared twice
5 | |         #[pin]
6 | |         field: &'__pin mut T,
7 | |     }
8 | | }
  | |_- previous declaration here

error[E0263]: lifetime name `'__pin` declared twice in the same scope
 --> $DIR/overlapping_lifetime_names.rs:4:20
  |
3 | / pin_project! { //~ ERROR E0496
4 | |     pub struct Foo<'__pin, T> { //~ ERROR E0263
  | |                    ^^^^^^ declared twice
5 | |         #[pin]
6 | |         field: &'__pin mut T,
7 | |     }
8 | | }
  | |_- previous declaration here
