error: no rules expected the token `[`
 --> $DIR/unsupported.rs:3:1
  |
3 | / pin_project! {
4 | |     struct Struct1 {} //~ ERROR no rules expected the token `}`
5 | | }
  | |_^ no rules expected this token in macro call
  |
  = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: no rules expected the token `[`
 --> $DIR/unsupported.rs:7:1
  |
7 | / pin_project! {
8 | |     struct Struct2(); //~ ERROR no rules expected the token `(`
9 | | }
  | |_^ no rules expected this token in macro call
  |
  = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: no rules expected the token `[`
  --> $DIR/unsupported.rs:11:1
   |
11 | / pin_project! {
12 | |     struct Struct3; //~ ERROR no rules expected the token `;`
13 | | }
   | |_^ no rules expected this token in macro call
   |
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: no rules expected the token `[`
  --> $DIR/unsupported.rs:15:1
   |
15 | / pin_project! {
16 | |     enum Enum { //~ ERROR no rules expected the token `enum`
17 | |         A(u8)
18 | |     }
19 | | }
   | |_^ no rules expected this token in macro call
   |
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: no rules expected the token `[`
  --> $DIR/unsupported.rs:21:1
   |
21 | / pin_project! {
22 | |     union Union { //~ ERROR no rules expected the token `union`
23 | |         x: u8,
24 | |     }
25 | | }
   | |_^ no rules expected this token in macro call
   |
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
