diff --git a/arm/arm_init.c b/arm/arm_init.c
--- a/arm/arm_init.c
+++ b/arm/arm_init.c
@@ -34,9 +34,9 @@
  * implementation which reads /proc/cpufino.
  */
 #ifndef PNG_ARM_NEON_FILE
 #  ifdef __linux__
-#     define PNG_ARM_NEON_FILE "contrib/arm-neon/linux.c"
+#     define PNG_ARM_NEON_FILE "linux.c"
 #  endif
 #endif
 
 #ifdef PNG_ARM_NEON_FILE
diff --git a/arm/filter_neon.S b/arm/filter_neon.S
--- a/arm/filter_neon.S
+++ b/arm/filter_neon.S
@@ -9,8 +9,14 @@
  * For conditions of distribution and use, see the disclaimer
  * and license in png.h
  */
 
+/* These are required because Mozilla's moz.build system doesn't pass
+ * -DDefined macros to the assembler.
+ */
+#define PNG_READ_SUPPORTED
+#define MOZ_PNG_HAVE_ARM_NEON
+
 /* This is required to get the symbol renames, which are #defines, and the
  * definitions (or not) of PNG_ARM_NEON_OPT and PNG_ARM_NEON_IMPLEMENTATION.
  */
 #define PNG_VERSION_INFO_ONLY
