From: Jeff Gilbert <jgilbert@mozilla.com>
Bug 724476 - ANGLE Bug 293 - Enforce readPixels format/type semantics

diff --git a/gfx/angle/src/libGLESv2/libGLESv2.cpp b/gfx/angle/src/libGLESv2/libGLESv2.cpp
--- a/gfx/angle/src/libGLESv2/libGLESv2.cpp
+++ b/gfx/angle/src/libGLESv2/libGLESv2.cpp
@@ -98,27 +98,16 @@ bool validReadFormatType(GLenum format, 
         switch (type)
         {
           case GL_UNSIGNED_BYTE:
             break;
           default:
             return false;
         }
         break;
-      case GL_BGRA_EXT:
-        switch (type)
-        {
-          case GL_UNSIGNED_BYTE:
-          case GL_UNSIGNED_SHORT_4_4_4_4_REV_EXT:
-          case GL_UNSIGNED_SHORT_1_5_5_5_REV_EXT:
-            break;
-          default:
-            return false;
-        }
-        break;
       case gl::IMPLEMENTATION_COLOR_READ_FORMAT:
         switch (type)
         {
           case gl::IMPLEMENTATION_COLOR_READ_TYPE:
             break;
           default:
             return false;
         }
