From: Jeff Gilbert <jgilbert@mozilla.com>

diff --git a/gfx/angle/src/libGLESv2/renderer/Renderer.cpp b/gfx/angle/src/libGLESv2/renderer/Renderer.cpp
--- a/gfx/angle/src/libGLESv2/renderer/Renderer.cpp
+++ b/gfx/angle/src/libGLESv2/renderer/Renderer.cpp
@@ -15,16 +15,23 @@
 #include "libGLESv2/renderer/Renderer11.h"
 #include "libGLESv2/utilities.h"
 
 #if !defined(ANGLE_ENABLE_D3D11)
 // Enables use of the Direct3D 11 API for a default display, when available
 #define ANGLE_ENABLE_D3D11 0
 #endif
 
+#define ANGLE_PRELOADED_D3DCOMPILER_MODULE_NAMES \
+    {                                            \
+        TEXT("d3dcompiler_46.dll"),              \
+        TEXT("d3dcompiler_43.dll")               \
+    }
+
+
 namespace rx
 {
 
 Renderer::Renderer(egl::Display *display) : mDisplay(display)
 {
     mD3dCompilerModule = NULL;
     mD3DCompileFunc = NULL;
 }
@@ -212,9 +219,9 @@ rx::Renderer *glCreateRenderer(egl::Disp
     return NULL;
 }
 
 void glDestroyRenderer(rx::Renderer *renderer)
 {
     delete renderer;
 }
 
-}
\ No newline at end of file
+}
