From: Jeff Gilbert <jgilbert@mozilla.com>

diff --git a/gfx/angle/include/GLSLANG/ShaderLang.h b/gfx/angle/include/GLSLANG/ShaderLang.h
--- a/gfx/angle/include/GLSLANG/ShaderLang.h
+++ b/gfx/angle/include/GLSLANG/ShaderLang.h
@@ -18,17 +18,17 @@
 #else  // defined(WIN32)
 #define COMPILER_EXPORT __attribute__((visibility("default")))
 #endif
 
 #else  // defined(COMPONENT_BUILD)
 #define COMPILER_EXPORT
 #endif
 
-#include "KHR/khrplatform.h"
+#include "khrplatform.h"
 #include <stddef.h>
 
 //
 // This is the platform independent interface between an OGL driver
 // and the shading language compiler.
 //
 
 #ifdef __cplusplus
