# HG changeset patch
# Parent 3f4e2a35c49bb865018798a76b62c169d8e3537b
diff --git a/gfx/angle/src/compiler/osinclude.h b/gfx/angle/src/compiler/osinclude.h
--- a/gfx/angle/src/compiler/osinclude.h
+++ b/gfx/angle/src/compiler/osinclude.h
@@ -7,17 +7,19 @@
 #ifndef __OSINCLUDE_H
 #define __OSINCLUDE_H
 
 //
 // This file contains contains os-specific datatypes and
 // declares any os-specific functions.
 //
 
-#if defined(_WIN32) || defined(_WIN64)
+#if defined(ANGLE_USE_NSPR)
+/* no need to define anything when using NSPR */
+#elif defined(_WIN32) || defined(_WIN64)
 #define ANGLE_OS_WIN
 #elif defined(__APPLE__) || defined(__linux__) || \
       defined(__FreeBSD__) || defined(__OpenBSD__) || \
       defined(__sun)
 #define ANGLE_OS_POSIX
 #else
 #error Unsupported platform.
 #endif
