From: Jeff Gilbert <jgilbert@mozilla.com>

diff --git a/gfx/angle/src/libGLESv2/renderer/Renderer9.cpp b/gfx/angle/src/libGLESv2/renderer/Renderer9.cpp
--- a/gfx/angle/src/libGLESv2/renderer/Renderer9.cpp
+++ b/gfx/angle/src/libGLESv2/renderer/Renderer9.cpp
@@ -3128,17 +3128,17 @@ D3DPOOL Renderer9::getTexturePool(DWORD 
     if (mD3d9Ex != NULL)
     {
         return D3DPOOL_DEFAULT;
     }
     else
     {
         if (!(usage & (D3DUSAGE_DEPTHSTENCIL | D3DUSAGE_RENDERTARGET)))
         {
-            return D3DPOOL_MANAGED;
+            return D3DPOOL_DEFAULT;
         }
     }
 
     return D3DPOOL_DEFAULT;
 }
 
 bool Renderer9::copyToRenderTarget(IDirect3DSurface9 *dest, IDirect3DSurface9 *source, bool fromManaged)
 {
