Only in ../_obj/make-3.81.90/config: Makefile
diff -u4 -r config/config.guess ../_obj/make-3.81.90/config/config.guess
--- config/config.guess	Fri Mar  9 09:32:15 2007
+++ ../_obj/make-3.81.90/config/config.guess	Fri Mar  9 09:56:53 2007
@@ -783,8 +783,11 @@
 	exit ;;
     *:MINGW*:*)
 	echo ${UNAME_MACHINE}-pc-mingw32
 	exit ;;
+    *:MSYS*:*)
+        echo ${UNAME_MACHINE}-pc-msys32
+        exit ;;
     i*:windows32*:*)
     	# uname -m includes "-pc" on this system.
     	echo ${UNAME_MACHINE}-mingw32
 	exit ;;
Only in ../_obj/make-3.81.90/config: config.guess~
diff -u4 -r config/config.sub ../_obj/make-3.81.90/config/config.sub
--- config/config.sub	Fri Mar  9 09:32:15 2007
+++ ../_obj/make-3.81.90/config/config.sub	Fri Mar  9 09:57:38 2007
@@ -682,8 +682,12 @@
 	mingw32)
 		basic_machine=i386-pc
 		os=-mingw32
 		;;
+        msys32)
+                basic_machine=i386-pc
+                os=-msys32
+                ;;
 	miniframe)
 		basic_machine=m68000-convergent
 		;;
 	*mint | -mint[0-9]* | *MiNT | *MiNT[0-9]*)
@@ -1213,9 +1217,9 @@
 	      | -ptx* | -coff* | -ecoff* | -winnt* | -domain* | -vsta* \
 	      | -udi* | -eabi* | -lites* | -ieee* | -go32* | -aux* \
 	      | -chorusos* | -chorusrdb* \
 	      | -cygwin* | -pe* | -psos* | -moss* | -proelf* | -rtems* \
-	      | -mingw32* | -linux-gnu* | -linux-newlib* | -linux-uclibc* \
+	      | -mingw32* | -msys32* | -linux-gnu* | -linux-newlib* | -linux-uclibc* \
 	      | -uxpv* | -beos* | -mpeix* | -udk* \
 	      | -interix* | -uwin* | -mks* | -rhapsody* | -darwin* | -opened* \
 	      | -openstep* | -oskit* | -conix* | -pw32* | -nonstopux* \
 	      | -storm-chaos* | -tops10* | -tenex* | -tops20* | -its* \
Only in ../_obj/make-3.81.90/config: config.sub~
