From 08125d302f8cc0bf000828fee92c7fd22523046f Mon Sep 17 00:00:00 2001
From: Tom Ritter <tom@ritter.vg>
Date: Tue, 7 Feb 2023 15:10:14 -0500
Subject: [PATCH 9/9] Update MFT_MESSAGE_TYPE

---
 mingw-w64-headers/include/mftransform.h | 23 +++++++++++++++--------
 1 file changed, 15 insertions(+), 8 deletions(-)

diff --git a/mingw-w64-headers/include/mftransform.h b/mingw-w64-headers/include/mftransform.h
index 663d2113a..50a3daac1 100644
--- a/mingw-w64-headers/include/mftransform.h
+++ b/mingw-w64-headers/include/mftransform.h
@@ -103,14 +103,21 @@ enum _MFT_PROCESS_OUTPUT_FLAGS {
     MFT_PROCESS_OUTPUT_REGENERATE_LAST_OUTPUT = 0x2
 };
 typedef enum _MFT_MESSAGE_TYPE {
-    MFT_MESSAGE_COMMAND_FLUSH = 0x0,
-    MFT_MESSAGE_COMMAND_DRAIN = 0x1,
-    MFT_MESSAGE_SET_D3D_MANAGER = 0x2,
-    MFT_MESSAGE_NOTIFY_BEGIN_STREAMING = 0x10000000,
-    MFT_MESSAGE_NOTIFY_END_STREAMING = 0x10000001,
-    MFT_MESSAGE_NOTIFY_END_OF_STREAM = 0x10000002,
-    MFT_MESSAGE_NOTIFY_START_OF_STREAM = 0x10000003,
-    MFT_MESSAGE_COMMAND_MARKER = 0x20000000
+    MFT_MESSAGE_COMMAND_FLUSH                   = 0x00000000,
+    MFT_MESSAGE_COMMAND_DRAIN                   = 0x00000001,
+    MFT_MESSAGE_SET_D3D_MANAGER                 = 0x00000002,
+    MFT_MESSAGE_DROP_SAMPLES                    = 0x00000003,
+    MFT_MESSAGE_COMMAND_TICK                    = 0x00000004,
+    MFT_MESSAGE_NOTIFY_BEGIN_STREAMING          = 0x10000000,
+    MFT_MESSAGE_NOTIFY_END_STREAMING            = 0x10000001,
+    MFT_MESSAGE_NOTIFY_END_OF_STREAM            = 0x10000002,
+    MFT_MESSAGE_NOTIFY_START_OF_STREAM          = 0x10000003,
+    MFT_MESSAGE_NOTIFY_RELEASE_RESOURCES        = 0x10000004,
+    MFT_MESSAGE_NOTIFY_REACQUIRE_RESOURCES      = 0x10000005,
+    MFT_MESSAGE_NOTIFY_EVENT                    = 0x10000006,
+    MFT_MESSAGE_COMMAND_SET_OUTPUT_STREAM_STATE = 0x10000007,
+    MFT_MESSAGE_COMMAND_FLUSH_OUTPUT_STREAM     = 0x10000008,
+    MFT_MESSAGE_COMMAND_MARKER                  = 0x20000000
 } MFT_MESSAGE_TYPE;
 typedef struct _MFT_INPUT_STREAM_INFO {
     LONGLONG hnsMaxLatency;
-- 
2.25.1

