error[E0277]: `PhantomPinned` cannot be unpinned
  --> tests/ui/not_unpin/assert-not-unpin.rs:31:5
   |
31 |     is_unpin::<Foo<(), ()>>(); //~ ERROR E0277
   |     ^^^^^^^^^^^^^^^^^^^^^^^ within `Wrapper<'_, PhantomPinned>`, the trait `Unpin` is not implemented for `PhantomPinned`
   |
   = note: consider using `Box::pin`
   = note: required because it appears within the type `Wrapper<'_, PhantomPinned>`
note: required because of the requirements on the impl of `Unpin` for `Foo<(), ()>`
  --> tests/ui/not_unpin/assert-not-unpin.rs:8:15
   |
8  | #[pin_project(!Unpin)]
   |               ^^^^^^
9  | struct Foo<T, U> {
   |        ^^^^^^^^^
note: required by a bound in `is_unpin`
  --> tests/ui/not_unpin/assert-not-unpin.rs:28:16
   |
28 | fn is_unpin<T: Unpin>() {}
   |                ^^^^^ required by this bound in `is_unpin`

error[E0277]: `PhantomPinned` cannot be unpinned
  --> tests/ui/not_unpin/assert-not-unpin.rs:32:5
   |
32 |     is_unpin::<Foo<PhantomPinned, ()>>(); //~ ERROR E0277
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ within `Wrapper<'_, PhantomPinned>`, the trait `Unpin` is not implemented for `PhantomPinned`
   |
   = note: consider using `Box::pin`
   = note: required because it appears within the type `Wrapper<'_, PhantomPinned>`
note: required because of the requirements on the impl of `Unpin` for `Foo<PhantomPinned, ()>`
  --> tests/ui/not_unpin/assert-not-unpin.rs:8:15
   |
8  | #[pin_project(!Unpin)]
   |               ^^^^^^
9  | struct Foo<T, U> {
   |        ^^^^^^^^^
note: required by a bound in `is_unpin`
  --> tests/ui/not_unpin/assert-not-unpin.rs:28:16
   |
28 | fn is_unpin<T: Unpin>() {}
   |                ^^^^^ required by this bound in `is_unpin`

error[E0277]: `PhantomPinned` cannot be unpinned
  --> tests/ui/not_unpin/assert-not-unpin.rs:33:5
   |
33 |     is_unpin::<Foo<(), PhantomPinned>>(); //~ ERROR E0277
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ within `Wrapper<'_, PhantomPinned>`, the trait `Unpin` is not implemented for `PhantomPinned`
   |
   = note: consider using `Box::pin`
   = note: required because it appears within the type `Wrapper<'_, PhantomPinned>`
note: required because of the requirements on the impl of `Unpin` for `Foo<(), PhantomPinned>`
  --> tests/ui/not_unpin/assert-not-unpin.rs:8:15
   |
8  | #[pin_project(!Unpin)]
   |               ^^^^^^
9  | struct Foo<T, U> {
   |        ^^^^^^^^^
note: required by a bound in `is_unpin`
  --> tests/ui/not_unpin/assert-not-unpin.rs:28:16
   |
28 | fn is_unpin<T: Unpin>() {}
   |                ^^^^^ required by this bound in `is_unpin`

error[E0277]: `PhantomPinned` cannot be unpinned
  --> tests/ui/not_unpin/assert-not-unpin.rs:34:5
   |
34 |     is_unpin::<Foo<PhantomPinned, PhantomPinned>>(); //~ ERROR E0277
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ within `Wrapper<'_, PhantomPinned>`, the trait `Unpin` is not implemented for `PhantomPinned`
   |
   = note: consider using `Box::pin`
   = note: required because it appears within the type `Wrapper<'_, PhantomPinned>`
note: required because of the requirements on the impl of `Unpin` for `Foo<PhantomPinned, PhantomPinned>`
  --> tests/ui/not_unpin/assert-not-unpin.rs:8:15
   |
8  | #[pin_project(!Unpin)]
   |               ^^^^^^
9  | struct Foo<T, U> {
   |        ^^^^^^^^^
note: required by a bound in `is_unpin`
  --> tests/ui/not_unpin/assert-not-unpin.rs:28:16
   |
28 | fn is_unpin<T: Unpin>() {}
   |                ^^^^^ required by this bound in `is_unpin`

error[E0277]: `PhantomPinned` cannot be unpinned
  --> tests/ui/not_unpin/assert-not-unpin.rs:36:5
   |
36 |     is_unpin::<TrivialBounds>(); //~ ERROR E0277
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^ within `Wrapper<'_, PhantomPinned>`, the trait `Unpin` is not implemented for `PhantomPinned`
   |
   = note: consider using `Box::pin`
   = note: required because it appears within the type `Wrapper<'_, PhantomPinned>`
note: required because of the requirements on the impl of `Unpin` for `TrivialBounds`
  --> tests/ui/not_unpin/assert-not-unpin.rs:15:15
   |
15 | #[pin_project(!Unpin)]
   |               ^^^^^^
16 | struct TrivialBounds {
   |        ^^^^^^^^^^^^^
note: required by a bound in `is_unpin`
  --> tests/ui/not_unpin/assert-not-unpin.rs:28:16
   |
28 | fn is_unpin<T: Unpin>() {}
   |                ^^^^^ required by this bound in `is_unpin`

error[E0277]: `PhantomPinned` cannot be unpinned
  --> tests/ui/not_unpin/assert-not-unpin.rs:38:5
   |
38 |     is_unpin::<Bar<'_, (), ()>>(); //~ ERROR E0277
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ within `Wrapper<'_, PhantomPinned>`, the trait `Unpin` is not implemented for `PhantomPinned`
   |
   = note: consider using `Box::pin`
   = note: required because it appears within the type `Wrapper<'_, PhantomPinned>`
note: required because of the requirements on the impl of `Unpin` for `Bar<'_, (), ()>`
  --> tests/ui/not_unpin/assert-not-unpin.rs:21:15
   |
21 | #[pin_project(!Unpin)]
   |               ^^^^^^
22 | struct Bar<'a, T, U> {
   |        ^^^^^^^^^^^^^
note: required by a bound in `is_unpin`
  --> tests/ui/not_unpin/assert-not-unpin.rs:28:16
   |
28 | fn is_unpin<T: Unpin>() {}
   |                ^^^^^ required by this bound in `is_unpin`

error[E0277]: `PhantomPinned` cannot be unpinned
  --> tests/ui/not_unpin/assert-not-unpin.rs:39:5
   |
39 |     is_unpin::<Bar<'_, PhantomPinned, PhantomPinned>>(); //~ ERROR E0277
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ within `Wrapper<'_, PhantomPinned>`, the trait `Unpin` is not implemented for `PhantomPinned`
   |
   = note: consider using `Box::pin`
   = note: required because it appears within the type `Wrapper<'_, PhantomPinned>`
note: required because of the requirements on the impl of `Unpin` for `Bar<'_, PhantomPinned, PhantomPinned>`
  --> tests/ui/not_unpin/assert-not-unpin.rs:21:15
   |
21 | #[pin_project(!Unpin)]
   |               ^^^^^^
22 | struct Bar<'a, T, U> {
   |        ^^^^^^^^^^^^^
note: required by a bound in `is_unpin`
  --> tests/ui/not_unpin/assert-not-unpin.rs:28:16
   |
28 | fn is_unpin<T: Unpin>() {}
   |                ^^^^^ required by this bound in `is_unpin`
