error[E0119]: conflicting implementations of trait `pin_project::UnsafeUnpin` for type `Foo<_, _>`:
  --> $DIR/impl-unsafe-unpin.rs:3:1
   |
3  | #[pin_project(!Unpin)] //~ ERROR E0119
   | ^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `Foo<_, _>`
...
10 | unsafe impl<T, U> UnsafeUnpin for Foo<T, U> where T: Unpin {}
   | ---------------------------------------------------------- first implementation here
   |
   = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0119]: conflicting implementations of trait `pin_project::UnsafeUnpin` for type `Bar<_, _>`:
  --> $DIR/impl-unsafe-unpin.rs:12:1
   |
12 | #[pin_project(!Unpin)] //~ ERROR E0119
   | ^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `Bar<_, _>`
...
19 | unsafe impl<T, U> UnsafeUnpin for Bar<T, U> {}
   | ------------------------------------------- first implementation here
   |
   = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0119]: conflicting implementations of trait `pin_project::UnsafeUnpin` for type `Baz<_, _>`:
  --> $DIR/impl-unsafe-unpin.rs:21:1
   |
21 | #[pin_project(!Unpin)] //~ ERROR E0119
   | ^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `Baz<_, _>`
...
28 | unsafe impl<T: Unpin, U: Unpin> UnsafeUnpin for Baz<T, U> {}
   | --------------------------------------------------------- first implementation here
   |
   = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)
