diff --git a/media/mp4parse-rust/mp4parse_capi/Cargo.toml b/media/mp4parse-rust/mp4parse_capi/Cargo.toml
index f04ad37ac568..a04b1e977735 100644
--- a/media/mp4parse-rust/mp4parse_capi/Cargo.toml
+++ b/media/mp4parse-rust/mp4parse_capi/Cargo.toml
@@ -18,6 +18,8 @@ exclude = [
   "*.mp4",
 ]
 
+build = false
+
 [badges]
 travis-ci = { repository = "https://github.com/mozilla/mp4parse-rust" }
 
@@ -30,9 +32,6 @@ num-traits = "0.2.0"
 [dev-dependencies]
 env_logger = "0.5.3"
 
-[build-dependencies]
-cbindgen = "0.5.2"
-
 [features]
 fuzz = ["mp4parse/fuzz"]
 # Enable mp4parse_fallible to use fallible memory allocation rather than
