diff --git a/media/mp4parse-rust/_upstream/mp4parse/mp4parse_capi/Cargo.toml b/media/mp4parse-rust/mp4parse_capi/Cargo.toml
index 2c606b4..895739f 100644
--- a/media/mp4parse-rust/_upstream/mp4parse/mp4parse_capi/Cargo.toml
+++ b/media/mp4parse-rust/mp4parse_capi/Cargo.toml
@@ -18,8 +18,7 @@ exclude = [
   "*.mp4",
 ]
 
-[lib]
-crate-type = ["lib", "cdylib"]
+build = false # See bug 1611431 - Generate mp4parse-rust bindings as part of mach build
 
 [badges]
 travis-ci = { repository = "https://github.com/mozilla/mp4parse-rust" }
@@ -33,9 +32,6 @@ num-traits = "0.2.0"
 [dev-dependencies]
 env_logger = "0.7.1"
 
-[build-dependencies]
-cbindgen = "0.5.2"
-
 [features]
 # Enable mp4parse_fallible to use fallible memory allocation rather than
 # panicking on OOM.  Note that this is only safe within Gecko where the system
