From db47e472d6f2393c58c1691d21e9bbc753b3d9da Mon Sep 17 00:00:00 2001
From: Mark Striemer <mstriemer@mozilla.com>
Date: Tue, 22 Nov 2022 18:20:46 -0600
Subject: [PATCH 4/5] Remove the bundled warning

---
 packages/lit/src/index.all.ts | 8 --------
 1 file changed, 8 deletions(-)

diff --git a/packages/lit/src/index.all.ts b/packages/lit/src/index.all.ts
index 1f8e5499..2bccd703 100644
--- a/packages/lit/src/index.all.ts
+++ b/packages/lit/src/index.all.ts
@@ -37,11 +37,3 @@ export {
   unsafeStatic,
   withStatic,
 } from './static-html.js';
-
-if (!window.litDisableBundleWarning) {
-  console.warn(
-    'Lit has been loaded from a bundle that combines all core features into ' +
-      'a single file. To reduce transfer size and parsing cost, consider ' +
-      'using the `lit` npm package directly in your project.'
-  );
-}
-- 
2.37.1 (Apple Git-137.1)

