--- lit.all.mjs
+++ lit.all.mjs.patched
@@ -358,8 +358,12 @@
     }
   }
   static createElement(t7, i9) {
-    const s8 = h2.createElement("template");
-    return s8.innerHTML = t7, s8;
+    const parser = new DOMParser();
+    const doc = parser.parseFromString(
+      `<template>${t7}</template>`,
+      "text/html"
+    );
+    return document.importNode(doc.querySelector("template"), true);
   }
 };
 function P(t7, i9, s8 = t7, e14) {
