--- a/fdlibm.h	2022-12-13 14:28:36.393051585 -0500
+++ b/fdlibm.h	2022-12-13 14:28:47.533092878 -0500
@@ -12,19 +12,21 @@
 /*
  * from: @(#)fdlibm.h 5.1 93/09/24
  * $FreeBSD$
  */
 
 #ifndef mozilla_imported_fdlibm_h
 #define	mozilla_imported_fdlibm_h
 
+namespace fdlibm {
+
 #ifndef M_PI_2
 #define	M_PI_2		1.57079632679489661923	/* pi/2 */
 #endif
 
 double	acos(double);
 double	asin(double);
 double	atan(double);
 double	atan2(double, double);
 double	cos(double);
 double	sin(double);
 double	tan(double);
@@ -69,9 +71,11 @@
 float	ceilf(float);
 float	fabsf(float);
 float	floorf(float);
 float	nearbyintf(float);
 float	rintf(float);
 float	scalbnf(float, int);
 float	truncf(float);
 
+} /* namespace fdlibm */
+
 #endif /* !mozilla_imported_fdlibm_h */
