diff --git a/modules/fdlibm/src/fdlibm.h b/modules/fdlibm/src/fdlibm.h
--- a/modules/fdlibm/src/fdlibm.h
+++ b/modules/fdlibm/src/fdlibm.h
@@ -19,16 +19,20 @@
 
 namespace fdlibm {
 
 double	acos(double);
 double	asin(double);
 double	atan(double);
 double	atan2(double, double);
 
+double	cos(double);
+double	sin(double);
+double	tan(double);
+
 double	cosh(double);
 double	sinh(double);
 double	tanh(double);
 
 double	exp(double);
 double	log(double);
 double	log10(double);
 
diff --git a/modules/fdlibm/src/math_private.h b/modules/fdlibm/src/math_private.h
--- a/modules/fdlibm/src/math_private.h
+++ b/modules/fdlibm/src/math_private.h
@@ -832,16 +832,19 @@ rnint(double_t x)
 #define	__ieee754_ynf	ynf
 #define	__ieee754_remainderf remainderf
 #define	__ieee754_scalbf scalbf
 
 #define acos fdlibm::acos
 #define asin fdlibm::asin
 #define atan fdlibm::atan
 #define atan2 fdlibm::atan2
+#define cos fdlibm::cos
+#define sin fdlibm::sin
+#define tan fdlibm::tan
 #define cosh fdlibm::cosh
 #define sinh fdlibm::sinh
 #define tanh fdlibm::tanh
 #define exp fdlibm::exp
 #define log fdlibm::log
 #define log10 fdlibm::log10
 #define pow fdlibm::pow
 #define ceil fdlibm::ceil
