diff --git a/media/omx-plugin/include/froyo/stagefright/DataSource.h b/media/omx-plugin/include/froyo/stagefright/DataSource.h
index 657f721..6f7dc38 100644
--- a/media/omx-plugin/include/froyo/stagefright/DataSource.h
+++ b/media/omx-plugin/include/froyo/stagefright/DataSource.h
@@ -26,19 +26,11 @@
 #include <utils/RefBase.h>
 #include <utils/threads.h>
 
-#if !defined(STAGEFRIGHT_EXPORT)
-#define STAGEFRIGHT_EXPORT
-#endif
-
-#if !defined(MOZ_STAGEFRIGHT_OFF_T)
-#define MOZ_STAGEFRIGHT_OFF_T off64_t
-#endif
-
 namespace android {
 
 class String8;
 
-class STAGEFRIGHT_EXPORT DataSource : public RefBase {
+class DataSource : public RefBase {
 public:
     enum Flags {
         kWantsPrefetching      = 1,
