diff --git a/media/libcubeb/src/cubeb.c b/media/libcubeb/src/cubeb.c
--- a/media/libcubeb/src/cubeb.c
+++ b/media/libcubeb/src/cubeb.c
@@ -227,24 +227,22 @@ cubeb_init(cubeb ** context, char const 
     wasapi_init,
 #endif
 #if defined(USE_WINMM)
     winmm_init,
 #endif
 #if defined(USE_SUN)
     sun_init,
 #endif
+#if defined(USE_AAUDIO)
+    aaudio_init,
+#endif
 #if defined(USE_OPENSL)
     opensl_init,
 #endif
-    // TODO: should probably be preferred over OpenSLES when available.
-    // Initialization will fail on old android devices.
-#if defined(USE_AAUDIO)
-    aaudio_init,
-#endif
 #if defined(USE_AUDIOTRACK)
     audiotrack_init,
 #endif
 #if defined(USE_KAI)
     kai_init,
 #endif
   };
   int i;
