diff --git a/media/libcubeb/src/cubeb_resampler.cpp b/media/libcubeb/src/cubeb_resampler.cpp
--- a/media/libcubeb/src/cubeb_resampler.cpp
+++ b/media/libcubeb/src/cubeb_resampler.cpp
@@ -50,18 +50,17 @@ passthrough_resampler<T>::passthrough_re
 
 template<typename T>
 long passthrough_resampler<T>::fill(void * input_buffer, long * input_frames_count,
                                     void * output_buffer, long output_frames)
 {
   if (input_buffer) {
     assert(input_frames_count);
   }
-  assert((input_buffer && output_buffer &&
-         *input_frames_count + static_cast<int>(samples_to_frames(internal_input_buffer.length())) >= output_frames) ||
+  assert((input_buffer && output_buffer) ||
          (output_buffer && !input_buffer && (!input_frames_count || *input_frames_count == 0)) ||
          (input_buffer && !output_buffer && output_frames == 0));
 
   if (input_buffer) {
     if (!output_buffer) {
       output_frames = *input_frames_count;
     }
     internal_input_buffer.push(static_cast<T*>(input_buffer),
