diff -ur libevent-1.4.7-stable/event.h libevent/event.h
--- libevent-1.4.7-stable/event.h	2008-05-28 23:43:46.000000000 -0700
+++ libevent/event.h	2008-09-11 16:05:42.000000000 -0700
@@ -159,7 +159,7 @@
 extern "C" {
 #endif
 
-#include <event-config.h>
+#include "event-config.h"
 #ifdef _EVENT_HAVE_SYS_TYPES_H
 #include <sys/types.h>
 #endif
@@ -172,7 +172,7 @@
 #include <stdarg.h>
 
 /* For int types. */
-#include <evutil.h>
+#include "evutil.h"
 
 #ifdef WIN32
 #define WIN32_LEAN_AND_MEAN
diff -ur libevent-1.4.7-stable/evutil.h libevent/evutil.h
--- libevent-1.4.7-stable/evutil.h	2008-06-27 17:49:53.000000000 -0700
+++ libevent/evutil.h	2008-09-11 16:06:28.000000000 -0700
@@ -38,7 +38,7 @@
 extern "C" {
 #endif
 
-#include <event-config.h>
+#include "event-config.h"
 #ifdef _EVENT_HAVE_SYS_TIME_H
 #include <sys/time.h>
 #endif
