diff --git a/intl/icu/source/tools/pkgdata/pkgdata.cpp b/intl/icu/source/tools/pkgdata/pkgdata.cpp
--- a/intl/icu/source/tools/pkgdata/pkgdata.cpp
+++ b/intl/icu/source/tools/pkgdata/pkgdata.cpp
@@ -2166,17 +2166,17 @@ static void pkg_createOptMatchArch(char 
     const char* obj = "oma.obj";
     FileStream* stream = NULL;
 
     stream = T_FileStream_open(source,"w");
     if (stream != NULL) {
         T_FileStream_writeLine(stream, code);
         T_FileStream_close(stream);
 
-        char cmd[SMALL_BUFFER_MAX_SIZE];
+        char cmd[LARGE_BUFFER_MAX_SIZE];
         sprintf(cmd, "%s %s -o %s",
             pkgDataFlags[COMPILER],
             source,
             obj);
 
         if (runCommand(cmd) == 0){
             sprintf(optMatchArch, "%s", obj);
         }
