error[E0277]: `PhantomPinned` cannot be unpinned
 --> $DIR/trivial_bounds-feature-gate.rs:8:5
  |
8 |     impl Unpin for A where PhantomPinned: Unpin {} //~ ERROR E0277
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Unpin` is not implemented for `PhantomPinned`
  |
  = help: see issue #48214
  = help: add `#![feature(trivial_bounds)]` to the crate attributes to enable

error[E0277]: `PhantomPinned` cannot be unpinned
   --> $DIR/trivial_bounds-feature-gate.rs:8:43
    |
8   |     impl Unpin for A where PhantomPinned: Unpin {} //~ ERROR E0277
    |                                           ^^^^^ the trait `Unpin` is not implemented for `PhantomPinned`

error[E0277]: `PhantomPinned` cannot be unpinned
  --> $DIR/trivial_bounds-feature-gate.rs:16:5
   |
16 |     impl Unpin for B where Wrapper<PhantomPinned>: Unpin {} //~ ERROR E0277
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Unpin` is not implemented for `PhantomPinned`
   |
   = note: required because of the requirements on the impl of `Unpin` for `phantom_pinned::Wrapper<PhantomPinned>`
   = help: see issue #48214
   = help: add `#![feature(trivial_bounds)]` to the crate attributes to enable

error[E0277]: `PhantomPinned` cannot be unpinned
  --> $DIR/trivial_bounds-feature-gate.rs:35:5
   |
35 |     impl Unpin for A where Inner: Unpin {} //~ ERROR E0277
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ within `Inner`, the trait `Unpin` is not implemented for `PhantomPinned`
   |
   = note: required because it appears within the type `Inner`
   = help: see issue #48214
   = help: add `#![feature(trivial_bounds)]` to the crate attributes to enable

error[E0277]: `PhantomPinned` cannot be unpinned
  --> $DIR/trivial_bounds-feature-gate.rs:43:5
   |
43 |     impl Unpin for B where Wrapper<Inner>: Unpin {} //~ ERROR E0277
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ within `Inner`, the trait `Unpin` is not implemented for `PhantomPinned`
   |
   = note: required because it appears within the type `Inner`
   = note: required because of the requirements on the impl of `Unpin` for `inner::Wrapper<Inner>`
   = help: see issue #48214
   = help: add `#![feature(trivial_bounds)]` to the crate attributes to enable
