error: trait bound PhantomPinned: Unpin does not depend on any type or lifetime parameters
  --> tests/ui/unstable-features/trivial_bounds.rs:15:43
   |
15 |     impl Unpin for A where PhantomPinned: Unpin {} //~ ERROR Unpin does not depend on any type or lifetime parameters
   |                                           ^^^^^
   |
note: the lint level is defined here
  --> tests/ui/unstable-features/trivial_bounds.rs:6:9
   |
6  | #![deny(trivial_bounds)]
   |         ^^^^^^^^^^^^^^

error: trait bound Inner: Unpin does not depend on any type or lifetime parameters
  --> tests/ui/unstable-features/trivial_bounds.rs:19:35
   |
19 |     impl Unpin for B where Inner: Unpin {} //~ ERROR Unpin does not depend on any type or lifetime parameters
   |                                   ^^^^^

error: trait bound Wrapper<Inner>: Unpin does not depend on any type or lifetime parameters
  --> tests/ui/unstable-features/trivial_bounds.rs:27:44
   |
27 |     impl Unpin for C where Wrapper<Inner>: Unpin {} //~ ERROR Unpin does not depend on any type or lifetime parameters
   |                                            ^^^^^
