diff --git a/aom_util/aom_pthread.h b/aom/aom_util/aom_pthread.h
--- a/aom_util/aom_pthread.h
+++ b/aom_util/aom_pthread.h
@@ -30,16 +30,18 @@ extern "C" {
 #define WIN32_LEAN_AND_MEAN
 #include <process.h>  // NOLINT
 #include <stddef.h>   // NOLINT
 #include <windows.h>  // NOLINT
 typedef HANDLE pthread_t;
 typedef int pthread_attr_t;
 typedef CRITICAL_SECTION pthread_mutex_t;
 
+#include <errno.h>
+
 #if _WIN32_WINNT < 0x0600
 #error _WIN32_WINNT must target Windows Vista / Server 2008 or newer.
 #endif
 typedef CONDITION_VARIABLE pthread_cond_t;
 
 #ifndef WINAPI_FAMILY_PARTITION
 #define WINAPI_PARTITION_DESKTOP 1
 #define WINAPI_FAMILY_PARTITION(x) x
