diff --git a/media/omx-plugin/include/ics/stagefright/DataSource.h b/media/omx-plugin/include/ics/stagefright/DataSource.h
index 81ef632..713af92 100644
--- a/media/omx-plugin/include/ics/stagefright/DataSource.h
+++ b/media/omx-plugin/include/ics/stagefright/DataSource.h
@@ -28,16 +28,12 @@
 #include <utils/threads.h>
 #include <drm/DrmManagerClient.h>
 
-#if !defined(STAGEFRIGHT_EXPORT)
-#define STAGEFRIGHT_EXPORT
-#endif
-
 namespace android {
 
 struct AMessage;
 class String8;
 
-class STAGEFRIGHT_EXPORT DataSource : public RefBase {
+class DataSource : public RefBase {
 public:
     enum Flags {
         kWantsPrefetching      = 1,
@@ -66,17 +62,6 @@ public:
         return 0;
     }
 
-#if defined(MOZ_ANDROID_SONY_WORKAROUND)
-private:
-    // Sony ICS devices have extra virtual methods in DataSource.
-    // The dummy methods below fill these vtable positions.
-    virtual void foo1() { abort(); }
-    virtual void foo2() { abort(); }
-    virtual void foo3() { abort(); }
-    virtual void foo4() { abort(); }
-public:
-#endif
-
     virtual status_t reconnectAtOffset(off64_t offset) {
         return ERROR_UNSUPPORTED;
     }

