diff --git a/media/libfishsound/src/libfishsound/fishsound_vorbis.c b/media/libfishsound/src/libfishsound/fishsound_vorbis.c
index 0c93a35..b1efc48 100644
--- a/media/libfishsound/src/libfishsound/fishsound_vorbis.c
+++ b/media/libfishsound/src/libfishsound/fishsound_vorbis.c
@@ -423,16 +423,18 @@ fs_vorbis_init (FishSound * fsound)
 
   fsv = fs_malloc (sizeof (FishSoundVorbisInfo));
   if (fsv == NULL) return NULL;
 
   fsv->packetno = 0;
   fsv->finished = 0;
   vorbis_info_init (&fsv->vi);
   vorbis_comment_init (&fsv->vc);
+  vorbis_dsp_init (&fsv->vd);
+  vorbis_block_init (&fsv->vd, &fsv->vb);
   fsv->pcm = NULL;
   fsv->ipcm = NULL;
   fsv->max_pcm = 0;
 
   fsound->codec_data = fsv;
 
 #if FS_ENCODE && HAVE_VORBISENC
 
