diff --git a/media/libfishsound/src/libfishsound/fishsound_comments.c b/media/libfishsound/src/libfishsound/fishsound_comments.c
index b71164f..414e210 100644
--- a/media/libfishsound/src/libfishsound/fishsound_comments.c
+++ b/media/libfishsound/src/libfishsound/fishsound_comments.c
@@ -290,7 +290,9 @@ fish_sound_comment_next_byname (FishSound * fsound,
 int
 fish_sound_comment_add (FishSound * fsound, FishSoundComment * comment)
 {
+#if FS_ENCODE
   FishSoundComment * new_comment;
+#endif
 
   if (fsound == NULL) return FISH_SOUND_ERR_BAD;
 
@@ -317,7 +319,9 @@ int
 fish_sound_comment_add_byname (FishSound * fsound, const char * name,
 			       const char * value)
 {
+#if FS_ENCODE
   FishSoundComment * comment;
+#endif
 
   if (fsound == NULL) return FISH_SOUND_ERR_BAD;
 
@@ -346,7 +350,9 @@ fish_sound_comment_add_byname (FishSound * fsound, const char * name,
 int
 fish_sound_comment_remove (FishSound * fsound, FishSoundComment * comment)
 {
+#if FS_ENCODE
   FishSoundComment * v_comment;
+#endif
 
   if (fsound == NULL) return FISH_SOUND_ERR_BAD;
 
@@ -372,8 +378,11 @@ fish_sound_comment_remove (FishSound * fsound, FishSoundComment * comment)
 int
 fish_sound_comment_remove_byname (FishSound * fsound, char * name)
 {
+#if FS_ENCODE
   FishSoundComment * comment;
-  int i, ret = 0;
+  int i;
+#endif
+  int ret = 0;
 
   if (fsound == NULL) return FISH_SOUND_ERR_BAD;
 
