diff --git a/modules/woff2/src/woff2_dec.cc b/modules/woff2/src/woff2_dec.cc
--- a/modules/woff2/src/woff2_dec.cc
+++ b/modules/woff2/src/woff2_dec.cc
@@ -22,16 +22,25 @@
 #include <cstring>
 #include <limits>
 #include <string>
 #include <vector>
 #include <map>
 #include <memory>
 #include <utility>
 
+#include "mozilla/UniquePtr.h"
+namespace std
+{
+  using mozilla::DefaultDelete;
+  using mozilla::UniquePtr;
+  #define default_delete DefaultDelete
+  #define unique_ptr UniquePtr
+}
+
 #include "./brotli/decode.h"
 #include "./buffer.h"
 #include "./port.h"
 #include "./round.h"
 #include "./store_bytes.h"
 #include "./table_tags.h"
 #include "./variable_length.h"
 #include "./woff2_common.h"
